// smui/ProgressBar/ProgressBar.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".progress-bar progress {\n  width: 160px;\n  height: 20px;\n  border: 1px solid #0064B4;\n  background-color: #e6e6e6;\n  color: #0064B4;\n  /*IE10*/\n}\n.progress-bar progress::-moz-progress-bar {\n  background: #0064B4;\n}\n.progress-bar progress::-webkit-progress-bar {\n  background: #FFF;\n}\n.progress-bar progress::-webkit-progress-value {\n  background: #0064B4;\n}\n.progress-bar .bar {\n  background-color: deepskyblue;\n  height: 10px;\n  border-radius: 5px;\n}\n.progress-bar .progress-box {\n  border: 1px solid #EEE;\n  border-radius: 5px;\n  background-color: #EEE;\n  width: 100%;\n  display: block;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    value: 0,
    height: 10,
    unit: '',
    showValue: false,
    type: 'custom',
    disabled: false,
    ratio: 0,
    max: 100
};

exports.default = {
    mixins: [_ui2.default],
    params: params,
    computed: {
        ratio: function ratio() {
            return Math.floor(this.value * 100 / this.max) + '%';
        }
    },
    methods: {
        changeValue: function changeValue(e) {
            if (this.disabled) {
                return false;
            }
            if (e) {
                var el = this.$el.querySelector('.progress-box');
                this.value = (e.offsetX * this.max / el.clientWidth).toFixed(2);
            }
            this.emit('change-value', { value: this.value });
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"progress-bar"},[(_vm.showValue)?_c('span',[_vm._v(_vm._s(_vm.value)+" "+_vm._s(_vm.unit))]):_vm._e(),_vm._v(" "),(_vm.type == 'custom')?_c('div',{staticClass:"progress-box",on:{"click":_vm.changeValue}},[_c('div',{staticClass:"bar",style:({width: _vm.ratio, height: _vm.height+'px'})})]):_vm._e(),_vm._v(" "),_c('progress',{directives:[{name:"show",rawName:"v-show",value:(_vm.type == 'native'),expression:"type == 'native'"}],attrs:{"value":_vm.value,"max":_vm.max},on:{"change":_vm.changeValue}})])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-44cf3805", __vue__options__)
  } else {
    hotAPI.reload("data-v-44cf3805", __vue__options__)
  }
})()}
