// smui/Pager/doc.js
'use strict';

exports.info = {
    name: 'Pager',
    desc: '翻页器',
    detail: ['前端翻页：传入数组自动分页', '后端翻页：每次请求后端更新翻页信息', '定制页面大小', '跳转指定页面']
};

exports.events = [{ name: 'page-change', data: '{currentPage, pageSize}', desc: '页号变化' }, { name: 'size-change', data: '{currentPage, pageSize}', desc: '大小变化' }, { name: 'list-change', data: '{list}', desc: '自动分页' }];

exports.methods = {
    setValue: {
        args: 'begin, end',
        desc: '设置取值'
    },
    ensure: {
        args: '',
        desc: '确定选择'
    }
};

exports.params = {
    needJumper: {
        default: true,
        type: ['Boolean'],
        desc: '是否显示跳页器'
    },
    sizeOptions: {
        default: '[20, 50, 100]',
        type: ['Array'],
        desc: '可选页大小'
    },
    currentPage: {
        default: 1,
        type: 'Number',
        desc: '当前页号'
    },
    pageSize: {
        default: 1,
        type: 'Number',
        desc: '当前页大小'
    },
    totalSize: {
        default: 1,
        type: 'Number',
        desc: '当前条目总数'
    },
    displayNum: {
        default: 6,
        type: 'Number',
        desc: '显示列数量'
    },
    items: {
        default: 1,
        type: 'Number',
        desc: '源数组（提供自动翻页）'
    }
};
