// smui/index.js
'use strict';

exports.DialogPanel = require('./DialogPanel');
exports.Calendar = require('./Calendar');
exports.DatePicker = require('./DatePicker');
exports.CityPicker = require('./CityPicker');
exports.DropList = require('./DropList');
exports.SelectList = require('./SelectList');
exports.FieldTip = require('./FieldTip');
exports.Info = require('./Info');
exports.Pager = require('./Pager');
exports.LineBox = require('./LineBox');
exports.NumberBox = require('./NumberBox');
exports.PressTab = require('./PressTab');
exports.RangeBar = require('./RangeBar');
exports.ProgressBar = require('./ProgressBar');
exports.Loading = require('./Dialog/Loading');
exports.Stepper = require('./Stepper');
exports.RichTable = require('./RichTable');
exports.Schedule = require('./Schedule');
exports.Scrum = require('./Scrum');
exports.SelectList = require('./SelectList');
exports.MultiSelect = require('./MultiSelect');
exports.Tab = require('./Tab');
exports.Tree = require('./Tree');
exports.Region = require('./Region');
exports.ScrollBar = require('./ScrollBar');
exports.Switch = require('./Switch');
exports.StarRate = require('./StarRate');

var Dialog = require('./Dialog');

window.SMUI = Object.assign({ Dialog: Dialog }, exports);

if (typeof window.define === 'function') {
    window.define('smui', window.SMUI);

    // 为了方便依赖，定义vue模块，可直接使用
    if (!window.Vue) {
        window.Vue = require('vue');
        window.define('vue', function () {
            return window.Vue;
        });
    }

    // 注册所有子模块
    for (var key in exports) {
        if (exports.hasOwnProperty(key)) {
            (function () {
                var comp = exports[key];
                window.define('smui/' + key, function () {
                    return comp;
                });
            })();
        }
    }
}

// 如果系统不支持require，自定义一个require机制
if (window.require === undefined) {
    var ui;

    (function () {
        var _modules = window._modules = window._modules || {};
        _modules['vue'] = window.Vue;
        for (ui in window.SMUI) {
            _modules['smui/' + ui] = window.SMUI[ui];
        }
        window.require = function (key) {
            if (key) {
                return _modules[key] || {};
            }
        };
    })();
}
