// smui/Draggable/Contianer.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".draggable-container {\n    position: relative;\n    .draggable {\n        cursor: move;\n    }\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    mixins: [_ui2.default],
    mounted: function mounted() {
        this.$el.classList.add('container');
        var els = this.$el.querySelectorAll('.draggable');
        els.forEach(function (el) {
            el.draggable = true;
        });
    },

    methods: {
        drag: function drag(ev) {
            var el = ev.target;
            var info = (0, _assign2.default)({}, el.dataset);
            info.ox = ev.offsetX;
            info.oy = ev.offsetY;
            info.id = el.id = el.id || 'drage-' + Math.floor(Math.random() * 10000);
            ev.dataTransfer.setData('dragInfo', (0, _stringify2.default)(info));
        },
        allowDrop: function allowDrop(ev) {
            var target = ev.target;
            while (target.parentElement && !target.draggable) {
                target = target.parentElement;
            }
            this.emit('dragmove', { id: target.id });
            ev.preventDefault();
        },
        drop: function drop(ev) {
            var info = ev.dataTransfer.getData('dragInfo');
            var target = ev.target;
            while (target.parentElement && !target.draggable) {
                target = target.parentElement;
            }
            if (info) {
                info = JSON.parse(info);
                if (target === this.$el) {
                    info.x = ev.offsetX - info.ox || 0;
                    info.y = ev.offsetY - info.oy || 0;
                } else {
                    var x = target.offsetLeft;
                    var y = target.offsetTop;
                    info.x = x + ev.offsetX - info.ox || 0;
                    info.y = y + ev.offsetY - info.oy;
                }
                this.emit('dropdown', info);
            }
            ev.preventDefault();
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"draggable-container",on:{"drop":function($event){_vm.drop($event)},"dragover":function($event){_vm.allowDrop($event)},"dragstart":function($event){_vm.drag($event)}}},[_vm._t("default")],2)}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-e48e0c52", __vue__options__)
  } else {
    hotAPI.reload("data-v-e48e0c52", __vue__options__)
  }
})()}
