// smui/Draggable/Draggable.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".smui.draggable {\n  border: 1px solid red;\n  position: absolute;\n  min-width: 50px;\n  min-height: 50px;\n  cursor: pointer;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = {
    props: ['x', 'y'],
    computed: {
        style: function style() {
            var i = this;
            return {
                left: i.x + 'px',
                top: i.y + 'px'
            };
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"draggable",style:(_vm.style)},[_vm._t("default")],2)}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-03e033d3", __vue__options__)
  } else {
    hotAPI.reload("data-v-03e033d3", __vue__options__)
  }
})()}
