// smui/base/i18n.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.i18n = exports.messages = undefined;

var _vueI18n = require('vue-i18n');

var _vueI18n2 = _interopRequireDefault(_vueI18n);

var _vue = require('vue');

var _vue2 = _interopRequireDefault(_vue);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// 避免Vue属于不同依赖
var Vue = window.Vue || _vue2.default;

// 全局侵入式增加翻译行数
Vue.prototype.$lt = function (key, pkg) {
    var flag = false;
    // 如果有本地方案
    if (this.$options.lang) {
        flag = this.$tr(key, pkg);
        if (flag !== undefined) {
            return flag;
        }
    }
    return i18n.t(key);
};

Vue.prototype.$tr = function (key, pkg) {
    var lang = this.$options.lang || pkg;
    if (!lang || !lang.hasOwnProperty(i18n.locale)) {
        return key;
    }
    return lang[i18n.locale][key];
};

Vue.use(_vueI18n2.default);

var messages = exports.messages = {
    en: {
        'ON': 'on',
        'OFF': 'off',
        'ENSURE': 'Ok',
        'CANCEL': 'Cancel',
        'CREATE': 'Create',
        'REMOVE': 'Remove',
        'DELETE': 'Delete',
        'UPDATE': 'Update',
        'MODIFY': 'Modify',
        'HELP_INFO': 'Help Info: ',
        'PLS_SELECT': 'SELECT PLS',
        'SELECTED': ' selected',
        'SELECT_ALL': 'All',
        'SELECT_ALT': 'Reverse',
        'LIMITED_DATES': 'Limited Dates: ',
        'RANGE_DATES': 'Range: ',
        'PAGE_SIZE': 'Page Size:',
        'JUMP_TO': 'Go to:',
        'CONFIG': 'Config',
        'HOUR': 'HOUR',
        'MINUTE': 'MIN',
        'SECOND': 'SEC'
    },
    cn: {
        'ON': '开',
        'OFF': '关',
        'ENSURE': '确定',
        'CANCEL': '取消',
        'CREATE': '创建',
        'REMOVE': '删除',
        'DELETE': '删除',
        'UPDATE': '更新',
        'MODIFY': '修改',
        'CONFIG': '配置',
        'PLS_SELECT': '请选择',
        'SELECTED': '项已选',
        'SELECT_ALL': '全选',
        'SELECT_ALT': '反选',
        'HELP_INFO': '说明：',
        'LIMITED_DATES': '限制天数：',
        'RANGE_DATES': '可选范围：',
        'PAGE_SIZE': '每页显示：',
        'JUMP_TO': '跳到：',
        'HOUR': '时',
        'MINUTE': '分',
        'SECOND': '秒'
    }
};

_vueI18n2.default.prototype.extend = function (pkg) {
    for (var key in pkg) {
        var lang = pkg[key];
        if (messages.hasOwnProperty(key)) {
            Object.assign(messages[key], lang);
        } else {
            messages[key] = lang;
        }
    }
};

var i18n = exports.i18n = new _vueI18n2.default({
    locale: 'cn',
    silentTranslationWarn: true,
    messages: messages
});

exports.default = i18n;
