// smui/base/util.js
'use strict';

var _vue = require('vue');

var _vue2 = _interopRequireDefault(_vue);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// 避免Vue属于不同依赖
var Vue = window.Vue || _vue2.default;

exports.extend = function (target, source) {
    Object.assign(target, source);
    return target;
};

exports.isEqual = function (obj, target) {
    return Object.is(obj, target);
};

exports.deepClone = function (obj) {
    return JSON.parse(JSON.stringify(obj));
};

exports.throttle = function (func, first, delay) {
    var timer = 0;
    var flag = true;
    if (typeof first !== 'function') {
        delay = +first;
        first = null;
    }
    return function () {
        var lastThis = this || {};
        var args = arguments;
        if (flag && first) {
            lastThis.scrolling = true;
            first.apply(lastThis, args);
            flag = false;
        }
        if (args.length && args[0] === true) {
            // 立即执行动作
            func.apply(lastThis, args);
        }
        clearTimeout(timer);
        timer = setTimeout(function () {
            func.apply(lastThis, args);
            lastThis.scrolling = false;
            flag = true;
        }, delay);
    };
};

exports.byteLength = function (str) {
    return str.replace(/[^\x00-\xff]/g, 'xx').length;
};

exports.wrapper = function (Comp) {
    return function (option) {
        option = option || {};
        var dom = document.createElement(option.modal ? 'dialog' : 'div');
        dom.class = 'dialog-instance';
        dom.id = 'dialog' + new Date().getTime();
        var container = document.body;
        if (option.container) {
            container = option.container;
            if (typeof option.container === 'string') {
                container = document.querySelector(option.container);
            }
        }
        container.appendChild(dom);
        option.extends = Comp;
        var inst = {};
        if (Vue.version.charAt(0) > 0) {
            inst = new Vue(option);
            inst.$mount(dom);
        } else {
            option.el = dom;
            inst = new Vue(option);
        }
        return inst;
    };
};
