// smui/Info/Info.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".info-sign {\n  position: relative;\n  line-height: 1.5em;\n  padding: 5px 0;\n  display: inline;\n}\n.info-sign .info {\n  width: 100px;\n  min-height: 50px;\n  box-sizing: border-box;\n  position: absolute;\n  background: #FFF;\n  display: block;\n  padding: 8px 16px;\n  border: 1px solid rgba(34, 36, 38, 0.2);\n  border-radius: 4px;\n  box-shadow: 0 2px 3px rgba(34, 36, 38, 0.15);\n  color: #333;\n  font-size: 13px;\n  font-weight: normal;\n  line-height: 1.8;\n  z-index: 999;\n}\n.info-sign .info:after,\n.info-sign .info:before {\n  content: '';\n  position: absolute;\n  border: 6px solid transparent;\n}\n.info-sign .info.left {\n  left: -10px;\n  top: -5px;\n  transform: translateX(-100%);\n}\n.info-sign .info.left:after,\n.info-sign .info.left:before {\n  right: -15px;\n  top: 10px;\n  margin-bottom: -20px;\n  border-left-width: 10px;\n}\n.info-sign .info.left:before {\n  margin-right: -2px;\n  border-left-color: rgba(34, 36, 38, 0.2);\n}\n.info-sign .info.left:after {\n  border-left-color: #FFF;\n}\n.info-sign .info.right {\n  left: 20px;\n  top: -5px;\n}\n.info-sign .info.right:after,\n.info-sign .info.right:before {\n  left: -15px;\n  top: 10px;\n  margin-bottom: -20px;\n  border-left-width: 10px;\n}\n.info-sign .info.right:before {\n  margin-left: -2px;\n  border-right-color: rgba(34, 36, 38, 0.2);\n}\n.info-sign .info.right:after {\n  border-right-color: #FFF;\n}\n.info-sign .info.bottom {\n  top: 32px;\n  left: 0;\n  transform: translateX(-50%);\n  min-width: 200px !important;\n}\n.info-sign .info.bottom:after,\n.info-sign .info.bottom:before {\n  content: '';\n  position: absolute;\n  left: 50%;\n  top: -15px;\n  border: 6px solid transparent;\n  border-bottom-width: 10px;\n}\n.info-sign .info.bottom:before {\n  margin-top: -2px;\n  border-bottom-color: rgba(34, 36, 38, 0.2);\n}\n.info-sign .info.bottom:after {\n  border-bottom-color: #FFF;\n}\n.info-sign .info.top {\n  top: 0;\n  left: 0;\n  transform: translateX(-50%) translateY(-100%);\n  min-width: 200px !important;\n}\n.info-sign .info.top:after,\n.info-sign .info.top:before {\n  content: '';\n  position: absolute;\n  left: 50%;\n  bottom: -15px;\n  border: 6px solid transparent;\n  border-top-width: 10px;\n}\n.info-sign .info.top:before {\n  margin-bottom: -2px;\n  border-top-color: rgba(34, 36, 38, 0.2);\n}\n.info-sign .info.top:after {\n  border-top-color: #FFF;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

var _layer = require('../base/layer');

var _layer2 = _interopRequireDefault(_layer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    isShow: false,
    hideOnLeave: true,
    direction: 'bottom',
    content: '信息'
};

exports.default = {
    params: params,
    mixins: [_ui2.default, _layer2.default],
    watch: {
        direction: function direction(value) {}
    },
    methods: {
        showInfo: function showInfo() {
            this.isShow = true;
        },
        hideInfo: function hideInfo() {
            if (this.hideOnLeave) {
                this.isShow = false;
            }
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"info-sign",on:{"mouseover":_vm.showInfo,"mouseout":_vm.hideInfo}},[_c('i',{staticClass:"sm-question",on:{"click":_vm.showInfo}}),_vm._v(" "),_c('div',{directives:[{name:"show",rawName:"v-show",value:(_vm.isShow),expression:"isShow"}],staticClass:"info",class:_vm.direction,style:(_vm.style)},[_vm._t("default",[_vm._v(_vm._s(_vm.content))])],2)])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-1c3e5812", __vue__options__)
  } else {
    hotAPI.reload("data-v-1c3e5812", __vue__options__)
  }
})()}
