// smui/MultiSelect/index.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".multi-select .layer.list {\n  width: 450px;\n  padding: 10px;\n}\n.multi-select .layer.list li {\n  display: inline-block;\n  min-width: 100px;\n}\n.multi-select .cutlist {\n  padding: 5px;\n}\n.multi-select .cutlist a:hover {\n  color: #08f18c;\n}\n.multi-select .opline {\n  min-height: 35px;\n  min-width: 80px;\n}\n.multi-select .opline .button {\n  margin: 5px 10px;\n  float: right;\n  height: 25px;\n  line-height: 25px;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _DropList = require('../DropList');

var _DropList2 = _interopRequireDefault(_DropList);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    value: [],
    oldValue: [],
    items: [],
    labels: [],
    title: 'SELECTED',
    isInputing: true,
    needCut: false,
    needEnsure: true,
    disabled: false,
    show: false
};

exports.default = {
    params: params,
    mixins: [_DropList2.default],
    computed: {
        itemMap: function itemMap() {
            var map = {};
            this.items.forEach(function (item) {
                map[item.value] = item;
            });
            return map;
        },
        selected: function selected() {
            var _this = this;

            if (this.value instanceof Array) {
                return this.value.map(function (val) {
                    return _this.itemMap[val];
                });
            }
            return [];
        }
    },
    mounted: function mounted() {
        this.oldValue = this.value;
        this.show = false;
    },

    methods: {
        reset: function reset() {
            if (this.needEnsure) {
                this.value = this.oldValue;
            } else {
                this.ensure(false);
            }
            this.show = false;
        },
        select: function select(keys) {
            var _this2 = this;

            if (keys === '*') {
                keys = (0, _keys2.default)(this.itemMap);
                this.value = keys;
            } else if (keys === '~') {
                keys = (0, _keys2.default)(this.itemMap);
                this.value = keys.filter(function (x) {
                    return !_this2.value.some(function (v) {
                        return x == v;
                    });
                });
            } else if (keys) {
                this.value = keys;
            }
            this.oldValue = this.value;
            this.emit('value-change', {
                value: this.value,
                selected: this.selected
            });
        },
        ensure: function ensure(isShow) {
            this.oldValue = this.value;
            var value = this.value;
            this.emit('value-change', {
                value: value,
                selected: this.selected
            });

            if (value && value.length) {
                this.title = '' + value.length + this.$lt('SELECTED');
            } else {
                this.title = this.params.title || 'PLS_SELECT';
            }
            this.show = isShow === true;
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"multi-select drop-list",class:{disabled: _vm.disabled},on:{"mouseleave":function($event){_vm.reset()}}},[_c('label',{staticClass:"title",attrs:{"disabled":_vm.disabled},on:{"click":function($event){_vm.show = !_vm.show}}},[_vm._v(_vm._s(_vm.$lt(_vm.title))+" \n        "),_c('i',{staticClass:"drop-icon sm-arrow-dropdown"})]),_vm._v(" "),_c('div',{directives:[{name:"show",rawName:"v-show",value:(_vm.show && !_vm.disabled),expression:"show && !disabled"}],class:{layer:1, list: _vm.items.length > 15}},[_c('ul',_vm._l((_vm.items),function(item){return _c('li',{attrs:{"value":item.value}},[_c('label',[_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.value),expression:"value"}],attrs:{"type":"checkbox","name":_vm.uid},domProps:{"value":item.value,"checked":Array.isArray(_vm.value)?_vm._i(_vm.value,item.value)>-1:(_vm.value)},on:{"change":function($event){var $$a=_vm.value,$$el=$event.target,$$c=$$el.checked?(true):(false);if(Array.isArray($$a)){var $$v=item.value,$$i=_vm._i($$a,$$v);if($$c){$$i<0&&(_vm.value=$$a.concat($$v))}else{$$i>-1&&(_vm.value=$$a.slice(0,$$i).concat($$a.slice($$i+1)))}}else{_vm.value=$$c}}}}),_vm._v("\n                    "+_vm._s(item.name)+"\n                ")])])})),_vm._v(" "),(_vm.needCut)?_c('div',{staticClass:"cutlist"},[_c('a',{on:{"click":function($event){_vm.select('*')}}},[_vm._v("\n            "+_vm._s(_vm.$lt('SELECT_ALL')))]),_vm._v(" "),_c('a',{on:{"click":function($event){_vm.select('~')}}},[_vm._v("\n            "+_vm._s(_vm.$lt('SELECT_ALT')))])]):_vm._e(),_vm._v(" "),(_vm.needEnsure)?_c('div',{staticClass:"opline"},[_c('a',{staticClass:"button",on:{"click":_vm.ensure}},[_vm._v(_vm._s(_vm.$lt('ENSURE')))])]):_vm._e()])])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-db5f7080", __vue__options__)
  } else {
    hotAPI.reload("data-v-db5f7080", __vue__options__)
  }
})()}
