// smui/RichTable/AffixTable.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".simple-table {\n  border: 1px solid #e4e4e4;\n  border-top: none;\n  border-right: none;\n  position: relative;\n  background-color: #FFF;\n  max-width: 100%;\n  box-sizing: border-box;\n}\n.simple-table .plugin-items {\n  display: inline-block;\n  position: absolute;\n  right: 5px;\n  top: 10px;\n}\n.simple-table .affix-table {\n  position: absolute;\n  z-index: 9;\n  top: 0;\n  left: 0;\n}\n.simple-table .affix-table td {\n  background: #FFF;\n}\n.simple-table .table-body.scroll-x {\n  max-height: 80vh;\n  overflow-x: hidden!important;\n}\n.simple-table .table-body.scroll-y {\n  max-height: 80vh;\n  overflow-y: scroll!important;\n}\n.simple-table thead.head-row {\n  width: 100%;\n  position: static;\n  background: #EEE;\n  border-bottom: 1px solid #EEE;\n}\n.simple-table table {\n  border-spacing: 0;\n  width: 100%;\n}\n.simple-table thead {\n  background-color: #f5f5f5;\n}\n.simple-table thead tr td {\n  line-height: 2em;\n  vertical-align: baseline;\n  background: #EEE;\n}\n.simple-table td {\n  color: #333;\n  font-size: 12px;\n  line-height: 1.5em;\n  border-top: 1px solid #e4e4e4;\n  border-right: 1px solid #e4e4e4;\n  padding: 8px 5px;\n  position: relative;\n  vertical-align: middle;\n}\n.simple-table table thead tr td {\n  background: #fff;\n  background-color: #fbfbfb;\n  font-weight: 700;\n  color: #656565;\n}\n.simple-table table tr:hover {\n  background-color: #f4fbf9;\n}\n.simple-table .pad-row {\n  height: 30px;\n}\n.simple-table .empty-box {\n  height: 50px;\n  text-align: center;\n  line-height: 50px;\n  background-color: #EEE;\n}\n.simple-table .summary-row {\n  background: #EEE;\n}\n.simple-table .summary-row .title {\n  font-weight: bold;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

var _TableHead = require('./TableHead');

var _TableHead2 = _interopRequireDefault(_TableHead);

var _FieldEditor = require('../FieldEditor');

var _FieldEditor2 = _interopRequireDefault(_FieldEditor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    cellTextMap: {},
    headTextMap: {},
    cellComp: {},
    selected: [],
    fields: [],
    sortState: {},
    items: [],
    info: {},
    singleFieldSort: true,
    scrollX: false,
    scrollY: false,
    minRows: 0,
    emptyTip: false,
    summary: {},
    action: {},
    syncParams: false,
    defaultOrder: '-down',
    summaryData: {},
    fieldWidth: {},
    selectable: true,
    checkAllPages: false,
    needCheckAllPages: false,
    headRow: ['origin'] };

var globalTimer = 0;

exports.default = {
    params: params,
    mixins: [_ui2.default],
    private: {
        scrollTop: 0,
        heights: [],
        scrollLeft: 0
    },
    ready: function ready() {
        this.fixHeadWidth();
        clearInterval(globalTimer);
        if (this.scrollX) {
            this.scrollTop = Math.max(15, this.scrollTop);
        }
        if (!this.params) {
            this.params = {
                fields: this.fields,
                headTextMap: this.headTextMap
            };
        }
    },

    watch: {
        selected: function selected(value) {
            var items = this.items;
            var list = [];
            if (value && value.length) {
                value = value.filter(function (idx) {
                    if (items[idx].$disabled !== true) {
                        list.push(items[idx]);
                        return true;
                    }
                    return false;
                });
            }
            this.emit('selected-changed', value);
            this.emit('selected-list-changed', list);

            var selectors = this.$el.querySelectorAll('.row-select');
            var _iteratorNormalCompletion = true;
            var _didIteratorError = false;
            var _iteratorError = undefined;

            try {
                for (var _iterator = (0, _getIterator3.default)(selectors), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                    var selector = _step.value;

                    if (selector && selector.__vue__) {
                        selector = selector.__vue__;
                        selector.setOptions({
                            selected: this.selected,
                            items: this.items
                        });
                    }
                }
            } catch (err) {
                _didIteratorError = true;
                _iteratorError = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion && _iterator.return) {
                        _iterator.return();
                    }
                } finally {
                    if (_didIteratorError) {
                        throw _iteratorError;
                    }
                }
            }
        },
        items: function items(value) {
            var _this = this;

            setTimeout(function () {
                _this.fixHeadWidth();
            }, 200);
        }
    },
    computed: {
        padRows: function padRows() {
            var pad = this.minRows - this.items.length;
            return Math.max(pad, 0);
        },
        sumData: function sumData() {
            var data = {
                cnt: 0,
                data: {}
            };
            var flag = true;
            var summaryData = this.summaryData;

            if (!summaryData || !this.fields || !this.summary) {
                return data;
            }
            if (!(this.summary.fields || []).length) {
                return data;
            }

            var fieldMap = {};
            this.summary.fields.forEach(function (x) {
                fieldMap[x] = 1;
            });

            this.fields.forEach(function (field, idx) {
                if (fieldMap.hasOwnProperty(field)) {
                    data.data[field] = summaryData[field];
                    flag = false;
                } else if (flag) {
                    data.cnt++;
                } else {
                    data.data[field] = '-';
                }
            });
            return data;
        }
    },
    components: {
        TableHead: _TableHead2.default,
        FieldEditor: _FieldEditor2.default
    },
    methods: {
        click: function click(event) {
            var option = event.target.dataset;
            var target = event.target.closest('td');
            this.emit('td-click', target);
            if (this.action && option.action) {
                var func = this.action[option.action] || this[option.action];
                if (typeof func === 'function') {
                    var row = target.dataset.row;
                    event.data = this.items[+row] || {};
                    event.field = target.dataset.col;
                    event.option = option;
                    func.call(this, event);
                }
            }
        },
        sortField: function sortField(e) {
            var field = e.field;
            var state = this.sortState;
            var st = state[field];
            if (st === '') {
                st = this.defaultOrder || '-up';
            }
            var asc = st === '-up';

            if (this.singleFieldSort) {
                for (var key in state) {
                    if (key !== field) {
                        state[key] = '';
                    }
                }
            }

            state[field] = asc ? '-down' : '-up';

            this.emit('sort-field', {
                field: field,
                asc: asc
            });
        },
        scrollHorizon: function scrollHorizon(e) {
            if (this.body) {
                var x = this.body.scrollLeft = e.target.scrollLeft;
                this.affix.scrollLeft = x;
            }
        },
        scrollVertical: function scrollVertical(e) {
            this.scrollTop = e.target.scrollTop;
            if (this.scrollX) {
                this.scrollTop = Math.max(15, this.scrollTop);
            }
        },
        changeCheck: function changeCheck(e) {
            if (e.checkAll) {
                this.selected = this.items.map(function (item, idx) {
                    return idx;
                });
                this.checkAllPages = e.checkAllPages;
            } else {
                this.checkAll = false;
                this.checkAllPages = false;
                this.selected = [];
            }
        },
        fixHeadWidth: function fixHeadWidth() {
            if (!this.$el.querySelector) {
                return;
            }
            var headRow = this.$el.querySelector('.affix-head colgroup');
            var header = this.$el.querySelector('.table-content .head-row');
            var rows = this.$el.querySelectorAll('.table-content .table-row');
            this.heights = [].slice.call(rows).map(function (r) {
                return r.clientHeight + 'px';
            });
            if (headRow && header) {
                if (!this.affix) {
                    this.affix = this.$el.querySelector('.affix-head');
                    this.body = this.$el.querySelector('.table-body');
                }
                var wx = header.parentElement.clientWidth;
                headRow.parentElement.style.width = wx + 'px';
                var headCells = headRow.querySelectorAll('col');
                var tdList = header.querySelectorAll('td');
                for (var i = 0; i < tdList.length; i++) {
                    if (i >= headCells.length) {
                        continue;
                    }
                    var width = tdList[i].clientWidth;
                    headCells[i].style.width = width + 'px';
                }
            }
        },
        setFields: function setFields(fields) {
            this.fields = fields;
            this.params.fields = fields;
        },
        setItems: function setItems(items) {
            this.items = items;
        },
        getSelected: function getSelected() {
            return this.selected || [];
        },
        setSelected: function setSelected(value) {
            this.selected = value;
        },
        cellFilter: function cellFilter(key, item) {
            var value = item[key];
            var cellTextMap = this.cellTextMap;
            if (cellTextMap && typeof cellTextMap[key] === 'function') {
                return item && cellTextMap[key].call(this, item[key], item, key);
            } else if (cellTextMap && cellTextMap.fieldFilter) {
                return item && cellTextMap.fieldFilter.call(this, item[key], item, key);
            } else {
                return value === null ? this.emptyText || '-' : item[key];
            }
        },
        trickCellPlugin: function trickCellPlugin(field, item) {
            if (this.cellPlugin && this.cellPlugin[field]) {
                return this.cellPlugin[field].call(this, field, item);
            }
            return '';
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"smui-table simple-table"},[_c('div',{on:{"click":_vm.click}},[(_vm.scrollX || _vm.scrollY)?_c('div',{directives:[{name:"show",rawName:"v-show",value:(_vm.scrollX || _vm.scrollTop > 10),expression:"scrollX || scrollTop > 10"}],staticClass:"affix-head",class:{'scroll-y': _vm.scrollY, 'scroll-x': _vm.scrollX},on:{"scroll":_vm.scrollHorizon}},[_c('table',[_c('colgroup',[(_vm.selectable)?_c('col'):_vm._e(),_vm._v(" "),_vm._l((_vm.fields),function(field){return _c('col')})],2),_vm._v(" "),_c("table-head",{ref:"affix-head",tag:"thead",attrs:{"params":_vm.params},on:{"check-change":_vm.changeCheck}})])]):_vm._e(),_vm._v(" "),_c('div',{staticClass:"table-body",class:{'scroll-y': _vm.scrollY, 'scroll-x': _vm.scrollX, scrolling: _vm.scrollTop > 10},on:{"scroll":_vm.scrollVertical}},[_c('div',{staticStyle:{"position":"relative"}},[_c('table',{staticClass:"affix-table"},[_c("table-head",{ref:"head",tag:"thead",attrs:{"params":_vm.params},on:{"check-change":_vm.changeCheck}}),_vm._v(" "),_vm._l((_vm.items),function(item,row){return _c('tr',{staticClass:"table-row"},[(_vm.selectable)?_c('td',{style:({height: _vm.heights[row]}),attrs:{"data-col":"select","state":_vm.cellFilter('select', item)}},[_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.selected),expression:"selected"}],attrs:{"type":"checkbox","disabled":item.$disabled},domProps:{"value":row,"checked":Array.isArray(_vm.selected)?_vm._i(_vm.selected,row)>-1:(_vm.selected)},on:{"change":function($event){var $$a=_vm.selected,$$el=$event.target,$$c=$$el.checked?(true):(false);if(Array.isArray($$a)){var $$v=row,$$i=_vm._i($$a,$$v);if($$c){$$i<0&&(_vm.selected=$$a.concat($$v))}else{$$i>-1&&(_vm.selected=$$a.slice(0,$$i).concat($$a.slice($$i+1)))}}else{_vm.selected=$$c}}}})]):_vm._e()])})],2),_vm._v(" "),_c('table',{staticClass:"table-content"},[_c("table-head",{ref:"head",tag:"thead",attrs:{"params":_vm.params},on:{"check-change":_vm.changeCheck}}),_vm._v(" "),_vm._l((_vm.items),function(item,row){return _c('tr',{staticClass:"table-row",class:{disabled: item.$disabled}},[(_vm.selectable)?_c('td',{attrs:{"data-col":"select","state":_vm.cellFilter('select', item)}},[_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.selected),expression:"selected"}],attrs:{"type":"checkbox","disabled":item.$disabled},domProps:{"value":row,"checked":Array.isArray(_vm.selected)?_vm._i(_vm.selected,row)>-1:(_vm.selected)},on:{"change":function($event){var $$a=_vm.selected,$$el=$event.target,$$c=$$el.checked?(true):(false);if(Array.isArray($$a)){var $$v=row,$$i=_vm._i($$a,$$v);if($$c){$$i<0&&(_vm.selected=$$a.concat($$v))}else{$$i>-1&&(_vm.selected=$$a.slice(0,$$i).concat($$a.slice($$i+1)))}}else{_vm.selected=$$c}}}})]):_vm._e(),_vm._v(" "),_vm._l((_vm.fields),function(field){return _c('td',{attrs:{"data-col":field,"data-row":row}},[(_vm.cellComp[field])?_c(_vm.cellComp[field].tag,{key:item.id || field + row,tag:"component",attrs:{"params":_vm.cellComp[field],"item":item,"model":item[field]},on:{"model-change":function (value) {item[field] = value}}}):_c('div',{domProps:{"innerHTML":_vm._s(_vm.cellFilter(field, item))}}),_vm._v(" "),_c('div',{staticClass:"plugin-items",domProps:{"innerHTML":_vm._s(_vm.trickCellPlugin(field, item))}})],1)})],2)}),_vm._v(" "),_vm._l((_vm.padRows),function(row){return _c('tr',{staticClass:"pad-row"},[(_vm.selectable)?_c('td',{attrs:{"data-col":"select"}}):_vm._e(),_vm._v(" "),_vm._l((_vm.fields),function(field){return _c('td',{attrs:{"data-col":field,"data-row":row}})})],2)}),_vm._v(" "),(_vm.sumData.cnt && _vm.items.length)?_c('tr',{staticClass:"summary-row"},[(_vm.selectable)?_c('td',{attrs:{"data-col":"check"}}):_vm._e(),_vm._v(" "),(_vm.sumData.cnt)?_c('td',{staticClass:"title",attrs:{"colspan":_vm.sumData.cnt}},[_vm._v(_vm._s(_vm.summary.title))]):_vm._e(),_vm._v(" "),_vm._l((_vm.sumData.data),function(value,field){return _c('td',{attrs:{"data-col":field}},[_c('div',{domProps:{"innerHTML":_vm._s(value)}})])})],2):_vm._e()],2)]),_vm._v(" "),(_vm.emptyTip && _vm.padRows === 0 && _vm.items.length === 0)?_c('div',{staticClass:"empty-box",domProps:{"innerHTML":_vm._s(_vm.emptyTip)}}):_vm._e()])])])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-5102d1a3", __vue__options__)
  } else {
    hotAPI.reload("data-v-5102d1a3", __vue__options__)
  }
})()}
