// smui/RichTable/RowSelect.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert("td.row-select {\n  min-width: 40px;\n}\ntd.row-select .cell {\n  z-index: 99;\n}\ntd.row-select .pined.pin-x {\n  background-color: #EEE;\n  max-width: 40px;\n}\ntd.row-select .select-flip {\n  position: absolute;\n  z-index: 100;\n  width: 100px;\n  height: 75px;\n  padding: 6px;\n  line-height: 2rem;\n  border: 1px solid #DDD;\n  top: -1px;\n  left: -1px;\n  border-radius: 2px;\n  box-shadow: 2px 2px #CCC;\n  background-color: #EEE;\n}\ntd.row-select .select-flip label {\n  width: auto!important;\n  display: block;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    selected: [],
    items: [],
    needCheckAllPages: false,
    selectHover: false,
    checkAllPages: false,
    indeterminate: false,
    checkThisPage: false,
    checkAll: false
};

exports.default = {
    mixins: [_ui2.default],
    params: params,
    watch: {
        indeterminate: function indeterminate(value) {
            var $all = this.$el.querySelectorAll('.select-all');
            var _iteratorNormalCompletion = true;
            var _didIteratorError = false;
            var _iteratorError = undefined;

            try {
                for (var _iterator = (0, _getIterator3.default)($all), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                    var item = _step.value;

                    item.checked = value === 2;
                    item.indeterminate = value === 1;
                }
            } catch (err) {
                _didIteratorError = true;
                _iteratorError = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion && _iterator.return) {
                        _iterator.return();
                    }
                } finally {
                    if (_didIteratorError) {
                        throw _iteratorError;
                    }
                }
            }
        },
        selected: function selected(value) {
            if (!this.selected) {
                return;
            }
            if (this.selected.length === 0) {
                this.indeterminate = 0;
                this.checkAll = false;
                this.checkThisPage = false;
                this.checkAllPages = false;
                return false;
            } else if (this.selected.length === this.items.length) {
                this.indeterminate = 2;
                this.checkAll = true;
                this.checkThisPage = !this.checkAllPages;
                return true;
            } else if (this.selected.length < this.items.length) {
                this.indeterminate = 1;
                this.checkAll = false;
                this.checkAllPages = false;
                this.checkThisPage = false;
                return false;
            }
        },
        checkAllPages: function checkAllPages(_checkAllPages) {
            if (_checkAllPages) {
                this.checkAll = true;
                this.checkThisPage = false;
            } else if (this.checkAll) {
                this.checkThisPage = true;
            }
            this.$update({ checkAllPages: _checkAllPages });
        },
        checkAll: function checkAll(_checkAll) {
            this.$update({ checkAll: _checkAll });
            if (_checkAll) {
                this.changeCheck();
            }
        }
    },
    methods: {
        changeSelectList: function changeSelectList(e) {
            var checked = !!e.target.checked;
            this.checkAllPages = false;
            this.checkAll = checked;
            if (!checked) {
                this.changeCheck();
            }
        },
        changeCheck: function changeCheck() {
            this.$dispatch('check-change', {
                checkAll: this.checkAll,
                checkAllPages: this.checkAllPages
            });
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('td',{staticClass:"row-select head-cell",attrs:{"data-col":"select"},on:{"mouseover":function($event){_vm.selectHover = true},"mouseout":function($event){_vm.selectHover = false}}},[_c('div',{staticClass:"cell pin-x"},[_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.checkAll),expression:"checkAll"}],staticClass:"select-all",attrs:{"type":"checkbox"},domProps:{"checked":_vm.checkAll,"checked":Array.isArray(_vm.checkAll)?_vm._i(_vm.checkAll,null)>-1:(_vm.checkAll)},on:{"change":[function($event){var $$a=_vm.checkAll,$$el=$event.target,$$c=$$el.checked?(true):(false);if(Array.isArray($$a)){var $$v=null,$$i=_vm._i($$a,$$v);if($$c){$$i<0&&(_vm.checkAll=$$a.concat($$v))}else{$$i>-1&&(_vm.checkAll=$$a.slice(0,$$i).concat($$a.slice($$i+1)))}}else{_vm.checkAll=$$c}},function($event){_vm.changeSelectList($event)}]}}),_vm._v(" "),(_vm.needCheckAllPages)?_c('div',{directives:[{name:"show",rawName:"v-show",value:(_vm.selectHover),expression:"selectHover"}],staticClass:"select-flip"},[_c('label',[_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.checkThisPage),expression:"checkThisPage"}],staticClass:"select-all",attrs:{"type":"checkbox"},domProps:{"checked":Array.isArray(_vm.checkThisPage)?_vm._i(_vm.checkThisPage,null)>-1:(_vm.checkThisPage)},on:{"change":[function($event){var $$a=_vm.checkThisPage,$$el=$event.target,$$c=$$el.checked?(true):(false);if(Array.isArray($$a)){var $$v=null,$$i=_vm._i($$a,$$v);if($$c){$$i<0&&(_vm.checkThisPage=$$a.concat($$v))}else{$$i>-1&&(_vm.checkThisPage=$$a.slice(0,$$i).concat($$a.slice($$i+1)))}}else{_vm.checkThisPage=$$c}},function($event){_vm.changeSelectList($event)}]}}),_vm._v("\n                选择本页\n            ")]),_vm._v(" "),_c('label',[_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.checkAllPages),expression:"checkAllPages"}],attrs:{"type":"checkbox"},domProps:{"checked":Array.isArray(_vm.checkAllPages)?_vm._i(_vm.checkAllPages,null)>-1:(_vm.checkAllPages)},on:{"change":function($event){var $$a=_vm.checkAllPages,$$el=$event.target,$$c=$$el.checked?(true):(false);if(Array.isArray($$a)){var $$v=null,$$i=_vm._i($$a,$$v);if($$c){$$i<0&&(_vm.checkAllPages=$$a.concat($$v))}else{$$i>-1&&(_vm.checkAllPages=$$a.slice(0,$$i).concat($$a.slice($$i+1)))}}else{_vm.checkAllPages=$$c}}}}),_vm._v(" 选全部\n            ")])]):_vm._e()])])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-c540097e", __vue__options__)
  } else {
    hotAPI.reload("data-v-c540097e", __vue__options__)
  }
})()}
