// smui/RichTable/TableHead.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".affix-head {\n  width: 100%;\n  background: #fbfbfb;\n  position: absolute;\n  top: 0;\n  left: -1px;\n  z-index: 10;\n}\n.affix-head table {\n  border: 1px solid #EEE;\n}\n.affix-head td {\n  border-top: none;\n}\n.affix-head tr td:last-child {\n  border-right: none;\n}\n.affix-head.scroll-x:hover {\n  padding-bottom: 1px;\n  overflow-x: scroll;\n}\n.table-head .sm-reorder-down,\n.table-head .sm-reorder-up {\n  color: orange;\n}\n.table-head .field.info-sign .layer {\n  min-width: 350px;\n  display: none;\n}\n.table-head .field.info-sign:hover .layer {\n  display: block;\n}\n.table-head .plugin-items i {\n  margin-right: 10px;\n}\n.smui-table ::-webkit-scrollbar-track-piece {\n  width: 10px;\n  background-color: rgba(155, 155, 155, 0.1);\n}\n.smui-table ::-webkit-scrollbar {\n  width: 5px;\n  background-color: rgba(255, 255, 255, 0.1);\n  height: 8px;\n  cursor: move;\n  position: absolute;\n  right: -5px;\n}\n.smui-table ::-webkit-scrollbar:after {\n  content: '';\n  width: 20px;\n  height: 20px;\n  background: red;\n}\n.smui-table .focus:hover ::-webkit-scrollbar {\n  background-color: rgba(55, 55, 55, 0.2);\n}\n.smui-table ::-webkit-scrollbar-thumb {\n  height: 50px;\n  background: rgba(55, 55, 55, 0.3);\n  cursor: pointer;\n}\n.smui-table ::-webkit-scrollbar-thumb:hover {\n  width: 10px;\n  background: #4e4e4e;\n  cursor: pointer;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

var _RowSelect = require('./RowSelect');

var _RowSelect2 = _interopRequireDefault(_RowSelect);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    fieldWidth: {},
    selectable: true,
    pinCell: {},
    headTextMap: {},
    headPlugin: {},
    sortState: {},
    fieldInfo: {},
    minWidth: '100px',
    fields: []
};

var innerPlugin = {
    sort: function sort(key, col) {
        var st = this.sortState[key] || '';
        return '<i class="sm-reorder' + st + '" data-action="sortField"></i>';
    },
    info: function info(key, col, fields) {
        var info = this.fieldInfo[key] || '';
        var direct = fields.length - col < 2 ? 'left' : 'bottom';
        return '<div class="info-sign field">\n                    <i class="sm-question"></i>\n                    <div class="info layer ' + direct + '">' + info + '</div>\n                </div>';
    }
};

exports.default = {
    mixins: [_ui2.default],
    params: params,
    components: {
        RowSelect: _RowSelect2.default
    },
    methods: {
        changeCheck: function changeCheck(e) {
            this.$emit('check-change', e);
        },
        headFilter: function headFilter(field) {
            if (this.headTextMap) {
                return this.$tr(this.headTextMap[field], this.lang) || field;
            } else {
                return field;
            }
        },
        trickHeadPlugin: function trickHeadPlugin(field, col) {
            var _this = this;

            var fields = this.fields;
            var plugins = [];
            if (this.fieldInfo[field]) {
                plugins.push(innerPlugin.info);
            }
            if (this.sortState.hasOwnProperty(field)) {
                plugins.push(innerPlugin.sort);
            }
            if (this.headPlugin[field]) {
                plugins.push(this.headPlugin[field]);
            }
            var result = plugins.map(function (plugin) {
                return plugin.call(_this, field, col, fields);
            });
            return result.join('');
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('thead',{staticClass:"head-row table-head"},[_c('tr',[(_vm.selectable)?_c("RowSelect",{tag:"td",attrs:{"params":_vm.params},on:{"check-change":_vm.changeCheck}}):_vm._e(),_vm._v(" "),_vm._l((_vm.fields),function(field,col){return _c('td',{staticClass:"head-cell",attrs:{"data-col":field}},[_c('div',{staticClass:"cell",class:{'pin-x': _vm.pinCell && _vm.pinCell[field]}},[_c('div',{style:({width: _vm.fieldWidth[field] || _vm.minWidth}),domProps:{"innerHTML":_vm._s(_vm.headFilter(field))}}),_vm._v(" "),_c('div',{staticClass:"plugin-items",domProps:{"innerHTML":_vm._s(_vm.trickHeadPlugin(field, col))}})])])})],2)])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-1322e699", __vue__options__)
  } else {
    hotAPI.reload("data-v-1322e699", __vue__options__)
  }
})()}
