// smui/PostForm/PostForm.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".post-form {\n  position: relative;\n}\n.post-form input,\n.post-form textarea {\n  outline: none;\n  border: 1px solid #bbb;\n}\n.post-form .on input:invalid {\n  border: 1px solid red;\n}\n.post-form .on textarea:valid {\n  border: 1px solid green;\n}\n.post-form .on textarea:invalid {\n  border: 1px solid red;\n}\n.post-form .on input:disabled {\n  border: 1px solid #CCC;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    params: {
        autoCheck: false,
        message: '',
        mod: 'add'
    },
    mixins: [_ui2.default],
    ready: function ready() {
        if (this.form) {
            return;
        }
        this.form = this.$el;
        if (this.$el.querySelector) {
            this.form = this.$el.querySelector('form');
            this.form.onsubmit = this.submit.bind(this);
        }

        if (this.mod === 'edit') {
            this.validate();
        }
    },

    methods: {
        validate: function validate(input) {
            var _this = this;

            var fields = this.form.querySelectorAll('.field');
            this.errorMessage = '';

            fields = Array.prototype.slice.call(fields);
            fields.forEach(function (field) {
                if (!field.clientHeight || field.dataset.ignore) {
                    return true;
                }
                field.validate && field.validate();
                if (!field.validity.valid && !_this.errorMessage) {
                    _this.errorMessage = field.validationMessage;

                    if (field.reportValidity) {
                        field.reportValidity();
                    }
                }
            });
            if (this.errorMessage) {
                return false;
            }
            return true;
        },
        sendQuery: function sendQuery() {},
        submit: function submit(e) {
            e.preventDefault();
            if (!this.validate()) {
                this.showError();
            } else {
                this.sendQuery();
            }
            return false;
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"post-form"},[_c('form',[_vm._t("default")],2),_vm._v(" "),_c('div',{staticClass:"error"},[_vm._v(_vm._s(_vm.message))])])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-cdafc792", __vue__options__)
  } else {
    hotAPI.reload("data-v-cdafc792", __vue__options__)
  }
})()}
