// smui/FieldBox/FieldBox.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".field-box-content input {\n  line-height: 20px;\n}\n.field-box-content .text {\n  width: 50%;\n  float: left;\n  margin: 2px 0;\n}\n.field-box-content .field-tip {\n  display: block;\n  position: static;\n}\n.field-box-content .error {\n  color: #C00;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

var _SelectList = require('../SelectList');

var _SelectList2 = _interopRequireDefault(_SelectList);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    params: {
        tip: '',
        min: 0,
        max: window.Infinity,
        step: 0.1,
        value: '',
        disabled: false,
        isInputing: true,
        error: '',
        autoCheck: true,
        type: 'text'
    },
    mixins: [_ui2.default],
    components: {
        SelectList: _SelectList2.default
    },
    updated: function updated() {
        var _this = this;

        if (this.autoCheck) {
            setTimeout(function () {
                _this.validate();
            }, 1000);
        }
    },

    methods: {
        ensure: function ensure() {
            if (this.validate()) {
                this.$emit('finished', { value: this.value });
            }
        },
        changeValue: function changeValue(e) {
            this.error = '';
            var target = e.target;
            if (target && target.validationMessage) {
                this.error = target.validationMessage;
                return;
            }
            if (this.validate(target)) {
                this.emit('value-change', { value: this.value });
            }
        },
        validate: function validate(target) {
            target = target || this.$el.querySelector('input');
            if (target && target.validationMessage) {
                this.error = target.validationMessage;
                return false;
            }
            return true;
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"field-box-content"},[(_vm.tip)?_c('div',[_vm._v(_vm._s(_vm.tip))]):_vm._e(),_vm._v(" "),(_vm.type=='text')?_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.value),expression:"value"}],attrs:{"type":"text","disabled":_vm.disabled,"maxlength":_vm.max,"minlength":_vm.min,"name":"content"},domProps:{"value":_vm._s(_vm.value)},on:{"change":_vm.changeValue,"input":function($event){if($event.target.composing){ return; }_vm.value=$event.target.value}}}):_vm._e(),_vm._v(" "),(_vm.type=='url')?_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.value),expression:"value"}],attrs:{"type":"url","disabled":_vm.disabled,"maxlength":_vm.max,"minlength":_vm.min,"name":"content"},domProps:{"value":_vm._s(_vm.value)},on:{"change":_vm.changeValue,"input":function($event){if($event.target.composing){ return; }_vm.value=$event.target.value}}}):_vm._e(),_vm._v(" "),(_vm.type=='email')?_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.value),expression:"value"}],attrs:{"type":"email","disabled":_vm.disabled,"maxlength":_vm.max,"minlength":_vm.min,"name":"content"},domProps:{"value":_vm._s(_vm.value)},on:{"change":_vm.changeValue,"input":function($event){if($event.target.composing){ return; }_vm.value=$event.target.value}}}):_vm._e(),_vm._v(" "),(_vm.type=='phone')?_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.value),expression:"value"}],attrs:{"type":"phone","disabled":_vm.disabled,"maxlength":_vm.max,"minlength":_vm.min,"name":"content"},domProps:{"value":_vm._s(_vm.value)},on:{"change":_vm.changeValue,"input":function($event){if($event.target.composing){ return; }_vm.value=$event.target.value}}}):_vm._e(),_vm._v(" "),(_vm.type=='number')?_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.value),expression:"value"}],attrs:{"type":"number","disabled":_vm.disabled,"max":_vm.max,"min":_vm.min,"step":_vm.step},domProps:{"value":_vm._s(_vm.value)},on:{"change":_vm.changeValue,"input":function($event){if($event.target.composing){ return; }_vm.value=_vm._n($event.target.value)},"blur":function($event){_vm.$forceUpdate()}}}):_vm._e(),_vm._v(" "),(_vm.type === 'select')?[_c('select-list',{directives:[{name:"model",rawName:"v-model",value:(_vm.value),expression:"value"}],attrs:{"params":_vm.params},domProps:{"value":(_vm.value)},on:{"input":function($event){_vm.value=$event}}})]:_vm._e(),_vm._v(" "),_c('div',{staticClass:"error"},[_vm._v(_vm._s(_vm.error))])],2)}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-5ea0f252", __vue__options__)
  } else {
    hotAPI.reload("data-v-5ea0f252", __vue__options__)
  }
})()}
