// smui/DialogPanel/DialogPanel.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".wild-dialog .dialog {\n  left: 30%;\n  top: 10%;\n}\n.dialog {\n  position: absolute;\n  min-width: 180px;\n  min-height: 100px;\n  padding: 0;\n  border: 1px solid rgba(34, 36, 38, 0.2);\n  box-sizing: border-box;\n  box-shadow: 0 2px 3px rgba(34, 36, 38, 0.15);\n  background-color: #FFF;\n  z-index: 80;\n}\n.dialog .header {\n  margin: 0;\n  font-size: 13px;\n  color: #333;\n  box-sizing: border-box;\n  height: 40px;\n  padding: 10px 38px 10px 16px;\n  border-bottom: 1px solid #e4e4e4;\n  background-color: #f5f5f5;\n  line-height: 20px;\n}\n.dialog .content {\n  padding: 28px;\n  display: block;\n  min-width: 120px;\n  min-height: 20px;\n}\n.dialog .pad-bottom {\n  position: relative;\n  padding-bottom: 30px;\n  min-height: 100%;\n}\n.dialog .footer {\n  padding: 8px 16px;\n  bottom: 0;\n  min-height: 40px;\n  position: absolute;\n  border-top: 1px solid #e4e4e4;\n  width: 100%;\n  box-sizing: border-box;\n}\n.dialog .footer .default-bottom {\n  text-align: right;\n  min-width: 100px;\n}\n.dialog .footer .button {\n  margin-left: 5px;\n  cursor: pointer;\n  min-width: 20px;\n  line-height: 24px;\n}\n.dialog .footer .cancel-button {\n  padding: 0 5px;\n  border: 1px solid #FFF;\n  min-width: 30px;\n  display: inline-block;\n  background: #fff;\n  outline-color: #fff;\n  background-image: none;\n  margin-left: 2px;\n  margin-right: 0px;\n}\n.dialog::backdrop {\n  background-color: rgba(0, 0, 0, 0.2);\n}\n.dialog .close {\n  position: absolute;\n  cursor: pointer;\n  right: 11px;\n  top: 12px;\n  font-size: 14px;\n  color: #333;\n  z-index: 3;\n}\n.dialog.scaled {\n  position: relative;\n  margin-bottom: 30px;\n}\n.dialog.scaled .footer {\n  position: absolute;\n}\n.dialog.fixed {\n  position: fixed;\n  z-index: 999;\n  left: 0;\n  top: 0;\n}\n.dialog.full {\n  position: fixed;\n  z-index: 999;\n  left: 0;\n  top: 0;\n  width: 100%;\n  height: 100vh;\n  overflow: auto;\n}\n.dialog.inline {\n  padding-bottom: 0;\n  min-width: 150px;\n}\n.dialog.inline .close {\n  right: 2px;\n  top: 2px;\n}\n.dialog.inline .footer {\n  display: none;\n}\n.dialog.inline .content {\n  min-height: 20px;\n  padding: 5px;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _create = require('babel-runtime/core-js/object/create');

var _create2 = _interopRequireDefault(_create);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

var _vue = require('vue');

var _vue2 = _interopRequireDefault(_vue);

var _DialogElement = require('./DialogElement');

var _DialogElement2 = _interopRequireDefault(_DialogElement);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Vue = window.Vue || _vue2.default;

var isNative = !!(document.createElement('dialog') || {}).showModal;

var params = {
    title: '',
    content: '',
    native: isNative,
    removeOnClose: true,
    ensureText: 'ENSURE',
    cancelText: 'CANCEL',
    needClose: true,
    needFooter: true,
    data: {},
    headerView: 'head',
    contentView: 'ContentView',
    footerView: 'foot',
    needCancel: true
};

function getContainer() {
    var dom = document.createElement('div');
    dom.id = 'dialog-instance';
    var panel = document.querySelector('#dialog-container-wrapper');
    if (!panel) {
        panel = document.createElement('div');
        panel.id = 'dialog-container-wrapper';
        document.body.append(panel);
    }
    panel.appendChild(dom);
    return dom;
}

exports.default = {
    params: params,
    private: {
        left: '',
        top: '',
        width: '',
        height: ''
    },
    replace: false,
    mixins: [_ui2.default],
    created: function created() {
        var content = this.$options.content;
        if ((typeof content === 'undefined' ? 'undefined' : (0, _typeof3.default)(content)) === 'object') {
            if (Vue.version.charAt(0) === '1') {
                content = Vue.extend(content);
            }
            this.$options.components.ContentView = content;
        } else if (typeof content === 'string') {
            this.content = content;
        }
    },
    ready: function ready() {
        this.$dialog = this.$refs.dialog;

        if (!this.$slots) {
            this.$slots = {};
        }
    },

    components: {
        DialogElement: _DialogElement2.default
    },
    methods: {
        appendTo: function appendTo(container) {
            container.appendChild(this.$el);
        },
        showModal: function showModal(option) {
            this.show(option, true);
        },
        set: function set(option) {
            if (typeof option === 'string') {
                option = {
                    content: option
                };
            }
            (0, _assign2.default)(this, option);
        },
        delegateEvent: function delegateEvent(target) {
            var $vm = this;

            $vm.view = target;
            target.$emit = function (type, e) {
                $vm.$emit(type, e);

                if ((typeof e === 'undefined' ? 'undefined' : (0, _typeof3.default)(e)) === 'object') {
                    var ex = (0, _create2.default)(e || {});
                    ex.type = type;
                    $vm.$emit('*', ex);
                }
                if (type === 'close') {
                    $vm.close();
                }
            };

            target.$dispatch = target.$emit;
        },
        show: function show(option, isModal) {
            this.set(option);

            if (!this.$el || !this.$el.parentElement) {
                this.$mount(getContainer(option));
            }

            if (isModal) {
                this.$refs.dialog.showModal();
            } else {
                this.$refs.dialog.show();
            }

            if (this.$refs.view) {
                this.delegateEvent(this.$refs.view);
            }
        },

        ensure: function ensure() {
            var view = this.$refs.view || {};
            var opts = view.$options || {};
            if (view.ensure) {
                view.ensure();
            } else if (opts.events && opts.events.ensure) {
                    opts.events.ensure.call(view);
                } else {
                    this.$emit('ensure');
                }
        },
        close: function close(option) {
            if (!this.removeOnClose) {
                this.hide();
                return;
            }
            if (option && option.isMute === true && this.$dialog.open) {
                this.$dialog.close();
                return;
            }
            this.$emit('closing');
            if (this.$dialog.open) {
                this.$dialog.close();
            }
            this.$emit('closed');
            this.$el.remove();
        },
        hide: function hide() {
            this.$dialog.hide();
            this.$emit('hide');
        }
    },
    computed: {
        comps: function comps() {
            var comps = {};
            for (var key in this.$options.components) {
                comps[key] = key;
            }
            return comps;
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"dialog-panel",class:_vm.clazz},[_c('dialog-element',{ref:"dialog",attrs:{"params":{width: _vm.width, height: _vm.height, top: _vm.top, left: _vm.left, native: _vm.native, }}},[(_vm.needClose)?_c('div',{staticClass:"close rotate sm-close",on:{"click":_vm.close}}):_vm._e(),_vm._v(" "),_c('div',{staticClass:"dialog-wrapper",class:{'pad-bottom': _vm.needFooter}},[(_vm.title || _vm.$slots.title)?_c('div',{staticClass:"header"},[_vm._t("title",[_c('div',{attrs:{"show":_vm.title},domProps:{"innerHTML":_vm._s(_vm.title)}})])],2):_vm._e(),_vm._v(" "),_c('div',{staticClass:"content"},[_vm._t("content",[(_vm.comps[_vm.contentView])?_c(_vm.contentView,{ref:"view",tag:"component",attrs:{"params":_vm.data}}):_c('div',{domProps:{"innerHTML":_vm._s(_vm.content)}})])],2),_vm._v(" "),(_vm.needFooter)?_c('div',{staticClass:"footer"},[_vm._t("footer",[_c('div',{staticClass:"default-bottom"},[_c('button',{staticClass:"button primary ensure-button",on:{"click":_vm.ensure}},[_vm._v(_vm._s(_vm.$lt(_vm.ensureText)))]),_vm._v(" "),(_vm.needCancel)?_c('button',{staticClass:"button plain cancel-button",on:{"click":_vm.close}},[_vm._v(_vm._s(_vm.$lt(_vm.cancelText)))]):_vm._e()])])],2):_vm._e()])])],1)}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-dafcf21e", __vue__options__)
  } else {
    hotAPI.reload("data-v-dafcf21e", __vue__options__)
  }
})()}
