// smui/SuggestList/index.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".suggest-list {\n  position: relative;\n  min-width: 100px;\n  display: inline-block;\n  padding-left: 0;\n  line-height: 28px;\n}\n.suggest-list .val-label {\n  border: 1px solid #EEE;\n  border-radius: 2px;\n  height: 28px;\n  background-color: #FFF;\n}\n.suggest-list input {\n  height: 28px;\n  line-height: 28px;\n  border: 0;\n  width: 100%;\n  padding: 0 5px;\n}\n.suggest-list .layer {\n  min-width: 100%;\n  position: absolute;\n  top: 27px;\n  left: 0;\n  background: #FFF;\n  z-index: 777;\n  border: 1px solid #DDD;\n}\n.suggest-list .layer .inner {\n  max-height: 300px;\n  overflow-y: scroll;\n}\n.suggest-list .layer li {\n  padding: 0 5px;\n  cursor: pointer;\n}\n.suggest-list .layer li:hover {\n  background-color: #EEE;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _DropList = require('../DropList');

var _DropList2 = _interopRequireDefault(_DropList);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    value: '',
    name: '',
    query: '',
    title: '输入查询',
    mod: 'select',
    items: [],
    isInputing: true,
    getList: null,
    disabled: false,
    show: false
};

exports.default = {
    params: params,
    mixins: [_DropList2.default],
    methods: {
        selectItem: function selectItem() {
            var target = window.event.target;
            var value = target.value,
                textContent = target.textContent,
                _target$name = target.name,
                name = _target$name === undefined ? textContent : _target$name;

            if (!name || !value) {
                return;
            }
            (0, _assign2.default)(this, { name: name, value: value });
            this.emit('value-change', { name: name, value: value });
        },
        changeList: function changeList(query) {
            var _this = this;

            this.$emit('query-change', query);

            if (!this.getList) {
                return;
            }
            var result = this.getList(query);
            if (result && typeof result.then === 'function') {
                result.then(function (data) {
                    _this.items = data;
                    _this.show = true;
                });
            } else {
                this.items = result;
                this.show = true;
            }
            this.$nextTick(function (e) {
                _this.fixPosition();
            });
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"command-list suggest-list",class:{disabled: _vm.disabled},on:{"mouseleave":function($event){_vm.show = false},"click":function($event){_vm.show = true}}},[_c('div',{staticClass:"val-label",on:{"mousehover":function($event){_vm.show = true}}},[_vm._v(_vm._s(_vm.name))]),_vm._v(" "),_c('div',{directives:[{name:"show",rawName:"v-show",value:(_vm.show),expression:"show"}],ref:"layer",staticClass:"layer",on:{"click":_vm.selectItem}},[_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.query),expression:"query"}],attrs:{"type":"text","placeholder":_vm.title},domProps:{"value":_vm._s(_vm.query)},on:{"keyup":function($event){_vm.changeList(_vm.query)},"input":function($event){if($event.target.composing){ return; }_vm.query=$event.target.value}}}),_vm._v(" "),_c('div',{staticClass:"inner"},[_vm._t("default",[(_vm.items && _vm.items.length > 0)?_c('ul',_vm._l((_vm.items),function(item){return _c('li',{attrs:{"value":item.value}},[_vm._v(_vm._s(item.name))])})):_vm._e()])],2)])])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-0b065e2d", __vue__options__)
  } else {
    hotAPI.reload("data-v-0b065e2d", __vue__options__)
  }
})()}
