// smui/Cascader/Cascader.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".cascader {\n  position: relative;\n}\n.cascader .value-label {\n  display: inline-block;\n  height: 30px;\n  line-height: 30px;\n  min-width: 6em;\n  padding: 0 10px;\n  border: 1px solid #CCC;\n  border-radius: 2px;\n}\n.cascader .layer {\n  position: absolute;\n  z-index: 7;\n  top: 28px;\n  left: 0;\n  background: #FFF;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

var _List = require('./List');

var _List2 = _interopRequireDefault(_List);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    items: [],
    isInputing: true,
    showLayer: true,
    value: 'cascader'
};

exports.default = {
    mixins: [_ui2.default],
    params: params,
    components: { CasList: _List2.default }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"cascader smui",on:{"mouseleave":function($event){_vm.showLayer = false}}},[_c('div',{staticClass:"value-label",on:{"click":function($event){_vm.showLayer = true}}},[_vm._v(_vm._s(_vm.value))]),_vm._v(" "),_c('div',{directives:[{name:"show",rawName:"v-show",value:(_vm.showLayer),expression:"showLayer"}],staticClass:"layer"},[_c('cas-list',{attrs:{"params":_vm.params}})],1)])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-1e906d77", __vue__options__)
  } else {
    hotAPI.reload("data-v-1e906d77", __vue__options__)
  }
})()}
