// smui/Cascader/List.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".cascader-list {\n  display: block;\n  width: 300px;\n  max-width: 500px;\n  padding-top: 10px;\n  border: 1px solid #CCC;\n}\n.cascader-list li:hover > .cascader-list {\n  display: block !important;\n}\n.cascader-list .cascader-list {\n  display: none;\n  position: absolute;\n  right: -300px;\n  top: 0;\n  background: #EEE;\n}\n.cascader-list ul {\n  list-style: none;\n}\n.cascader-list li {\n  position: relative;\n  line-height: 30px;\n}\n.cascader-list li .checked {\n  background: lightblue;\n}\n.cascader-list li .item-title {\n  padding: 0 10px;\n  cursor: pointer;\n}\n.cascader-list li .item-title:hover {\n  background: #CCC;\n}\n.cascader-list .right-icon {\n  position: absolute;\n  right: 0;\n  top: 8px;\n  transform: rotateZ(-90deg);\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    name: 'CasList',
    props: ['params'],
    data: function data() {
        var params = this.params;
        var items = params.items || [];
        var level = params.level || 0;
        var value = params.value;
        items = items.map(function (item) {
            item.fold = level > 0;
            item.checked = item.value === value[level];
            return (0, _extends3.default)({}, item);
        });
        return {
            items: items,
            value: value,
            level: level
        };
    },

    methods: {
        check: function check(item, idx) {
            if (item.children) {
                return;
            }
            this.value[this.level] = item.value;
            if (this.params.pnode) {
                this.value[this.level - 1] = this.params.pnode.value;
                this.$parent.check(this.params.pnode);
            }
        },
        uncheck: function uncheck(item) {}
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"cascader-list"},[_c('ul',_vm._l((_vm.items),function(item){return _c('li',[_c('div',{staticClass:"item-title",class:{checked: item.value === _vm.value[_vm.level]},on:{"click":function($event){_vm.check(item)}}},[_c('span',{domProps:{"innerHTML":_vm._s(item.label)}}),_vm._v(" "),(item.children)?_c('i',{staticClass:"right-icon sm-arrow-dropdown"}):_vm._e()]),_vm._v(" "),(item.children)?_c('cas-list',{attrs:{"params":{items: item.children, level: _vm.level + 1, value: _vm.value, pnode: item}}}):_vm._e()],1)}))])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-54051657", __vue__options__)
  } else {
    hotAPI.reload("data-v-54051657", __vue__options__)
  }
})()}
