// smui/Calendar/Picker.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".calendar-picker {\n  position: relative;\n}\n.calendar-picker .cal-layer {\n  position: absolute;\n  background-color: #FFF;\n  z-index: 77;\n  border: 1px solid #EEE;\n}\n.calendar-picker .cal-bottom {\n  padding: 5px 10px 5px 170px;\n  border-top: 1px solid #EEE;\n}\n.calendar-picker .calendar {\n  padding-bottom: 0;\n}\n.calendar-picker .calendar:before {\n  position: absolute;\n  left: 30px;\n  top: -10px;\n  content: \"\";\n  border: 5px solid rgba(0, 0, 0, 0);\n  border-bottom-color: #DEDEDE;\n}\n.calendar-picker .calendar:after {\n  position: absolute;\n  left: 30px;\n  top: -9px;\n  content: \"\";\n  border: 5px solid rgba(0, 0, 0, 0);\n  border-bottom-color: #FFF;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

var _Calendar = require('./Calendar');

var _Calendar2 = _interopRequireDefault(_Calendar);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    value: '',
    syncParams: true,
    buttonClass: ['button'],
    rawValue: null,
    needEnsure: true,
    isInputing: true,
    showLayer: false
};

exports.default = {
    params: params,
    mixins: [_ui2.default],
    components: {
        Calendar: _Calendar2.default
    },
    methods: {
        changeValue: function changeValue() {
            if (!this.needEnsure) {
                this.value = this.rawValue;
                this.emit('value-change', { value: this.value });
            }
        },
        hideLayer: function hideLayer() {
            this.showLayer = false;
            if (this.needEnsure) {
                this.rawValue = this.value;
            }
        },
        ensure: function ensure() {
            this.value = this.rawValue;
            this.hideLayer();
            this.emit('value-change', { value: this.value });
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"calendar-picker",on:{"mouseleave":function($event){_vm.hideLayer()}}},[_c('div',{class:_vm.buttonClass,on:{"click":function($event){_vm.showLayer = true}}},[_vm._t("info-text",[_vm._v(_vm._s(_vm.value))]),_vm._v(" "),_c('i',{staticClass:"sm-calender"})],2),_vm._v(" "),_c('div',{directives:[{name:"show",rawName:"v-show",value:(_vm.showLayer),expression:"showLayer"}],staticClass:"cal-layer"},[_c('calendar',{directives:[{name:"model",rawName:"v-model",value:(_vm.rawValue),expression:"rawValue"}],attrs:{"params":_vm.params},domProps:{"value":(_vm.rawValue)},on:{"value-change":_vm.changeValue,"input":function($event){_vm.rawValue=$event}}}),_vm._v(" "),(_vm.needEnsure)?_c('div',{staticClass:"cal-bottom"},[_c('div',{staticClass:"button",on:{"click":_vm.ensure}},[_vm._v(_vm._s(_vm.$lt("ENSURE")))])]):_vm._e()],1)])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-bfa631f2", __vue__options__)
  } else {
    hotAPI.reload("data-v-bfa631f2", __vue__options__)
  }
})()}
