// smui/Calendar/YearPicker.js
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

var _i18n = require('./i18n');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var NOW = new Date().getFullYear();
exports.default = {
    mixins: [_ui2.default],
    lang: _i18n.lang,
    data: function data() {
        var year = this.params.value;
        var years = this.getYears(year);
        return (0, _assign2.default)({
            now: NOW,
            value: year,
            years: years
        }, this.params);
    },

    watch: {
        'params.value': function paramsValue(year) {
            this.value = year;
            this.years = this.getYears(year);
        }
    },
    methods: {
        getStatus: function getStatus(year) {
            var value = this.value,
                now = this.now,
                minYear = this.minYear,
                maxYear = this.maxYear;

            return {
                disabled: year < minYear || year > maxYear,
                selected: value === year,
                now: now === year
            };
        },
        select: function select(idx) {
            this.emit('value-change', { value: idx });
        },
        getYears: function getYears(year) {
            return new Array(12).fill(year).map(function (val, idx) {
                return val - 6 + idx;
            });
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"year-picker month-picker"},_vm._l((_vm.years),function(year){return _c('div',{staticClass:"month-item",class:_vm.getStatus(year),on:{"click":function($event){_vm.select(year)}}},[_vm._v(_vm._s(year)+"\n    ")])}))}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-7c2d8aa4", __vue__options__)
  } else {
    hotAPI.reload("data-v-7c2d8aa4", __vue__options__)
  }
})()}
