// smui/Switch/Switch.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".switch-button {\n  border: 1px solid #2db7f5;\n  position: relative;\n  display: inline-block;\n  box-sizing: border-box;\n  width: 50px;\n  height: 22px;\n  line-height: 20px;\n  vertical-align: middle;\n  border-radius: 20px 20px;\n  border: 1px solid rgba(204, 204, 204, 0.41);\n  background-color: #ccc;\n  cursor: pointer;\n  -webkit-transition: all 0.3s cubic-bezier(0.78, 0.14, 0.15, 0.86);\n  transition: all 0.3s cubic-bezier(0.78, 0.14, 0.15, 0.86);\n  -webkit-user-select: none;\n  -moz-user-select: none;\n  -ms-user-select: none;\n  user-select: none;\n  padding: 0 5px;\n}\n.switch-button span {\n  float: right;\n  color: #333;\n}\n.switch-button.on {\n  background-color: #2db7f5;\n}\n.switch-button.on:after {\n  left: inherit;\n  right: 1px;\n}\n.switch-button.on span {\n  float: left;\n  color: #fdfdfd;\n}\n.switch-button:after {\n  position: absolute;\n  width: 18px;\n  height: 18px;\n  left: 2px;\n  top: 1px;\n  border-radius: 100%;\n  background-color: #fff;\n  content: \" \";\n  cursor: pointer;\n  -webkit-transition: left 0.3s cubic-bezier(0.78, 0.14, 0.15, 0.86);\n  transition: left 0.3s cubic-bezier(0.78, 0.14, 0.15, 0.86);\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    mixins: [_ui2.default],
    params: {
        value: true,
        stateMap: {
            'true': 'ON',
            'false': 'OFF'
        }
    },
    methods: {
        toggle: function toggle() {
            this.value = !this.value;
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"switch-button",class:{on: _vm.value},on:{"click":_vm.toggle}},[_c('span',[_vm._v(_vm._s(_vm.$lt(_vm.stateMap[_vm.value])))])])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-42a49612", __vue__options__)
  } else {
    hotAPI.reload("data-v-42a49612", __vue__options__)
  }
})()}
