// smui/DropList/DropList.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".drop-list {\n  position: relative;\n  display: inline-block;\n  min-width: 100px;\n  padding-right: 20px;\n  height: 30px;\n  vertical-align: top;\n  border: 1px solid #DDD;\n  border-radius: 2px;\n  cursor: pointer;\n  background-color: #FFF;\n  line-height: 30px;\n  padding-left: .5em;\n}\n.drop-list .title {\n  display: inline-block;\n  width: 100%;\n  height: 100%;\n  overflow: hidden;\n}\n.drop-list .drop-icon {\n  line-height: 30px;\n  position: absolute;\n  background: none;\n  pointer-events: none;\n  height: 25px;\n  right: 5px;\n  top: 0;\n}\n.drop-list .layer {\n  min-width: 100%;\n  position: absolute;\n  top: 28px;\n  max-height: 300px;\n  overflow-x: hidden;\n  overflow-y: auto;\n  left: 0;\n  background: #FFF;\n  z-index: 777;\n  border: 1px solid #DDD;\n}\n.drop-list .layer.select li.selected {\n  background-color: #EEE;\n  color: #39bd8b;\n}\n.drop-list li {\n  padding: 0 10px;\n  line-height: 2em;\n  position: relative;\n}\n.drop-list li:hover {\n  background-color: #EEE;\n  color: #39bd8b;\n}\n.drop-list:disabled {\n  color: #EEE;\n}\n.drop-list:disabled .layer {\n  display: none !important;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    value: '',
    name: '',
    items: [],
    isInputing: true,
    title: 'PLS_SELECT',
    mod: 'command',
    checkText: '-',
    disabled: false,
    show: false
};

exports.default = {
    params: params,
    mixins: [_ui2.default],
    watch: {
        items: function items(val) {
            var _this = this;

            this.$nextTick(function () {
                _this.fixPosition();
            });
        },
        show: function show(val) {
            var _this2 = this;

            this.$nextTick(function () {
                _this2.fixPosition();
            });
        }
    },
    methods: {
        getShowText: function getShowText(item) {
            return '' + item.name;
        },
        selectItem: function selectItem(e) {
            var target = e.target;
            if (target.tagName !== 'LI' && target.tagName !== 'A') {
                return;
            }
            this.name = target.textContent;
            if (this.mod === 'select') {
                this.title = this.name;
            }

            this.value = target.getAttribute('value');
            this.show = false;
            this.emit('select-change', {
                value: this.value,
                name: this.name
            });
        },
        fixPosition: function fixPosition() {
            var layer = this.$el.querySelector('.layer');
            var rect = layer.getBoundingClientRect();
            if (rect.height < rect.top && rect.bottom > window.innerHeight) {
                layer.style.top = 'inherit';
                layer.style.bottom = '27px';
            } else {
                layer.style.bottom = 'inherit';
                layer.style.top = '27px';
            }
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"drop-list",class:{disabled: _vm.disabled},on:{"mouseleave":function($event){_vm.show = false}}},[_c('label',{staticClass:"title",attrs:{"disabled":_vm.disabled},on:{"click":function($event){_vm.show = !_vm.show}}},[_vm._v(_vm._s(_vm.$lt(_vm.title))+" \n        "),_c('i',{staticClass:"drop-icon sm-arrow-dropdown"})]),_vm._v(" "),_c('div',{directives:[{name:"show",rawName:"v-show",value:(_vm.show && !_vm.disabled),expression:"show && !disabled"}],ref:"layer",staticClass:"layer",class:[_vm.mod],on:{"click":_vm.selectItem}},[_vm._t("default",[_c('ul',_vm._l((_vm.items),function(item){return _c('li',{class:{selected: item.value == _vm.value},attrs:{"value":item.value},domProps:{"innerHTML":_vm._s(_vm.getShowText(item))}})}))])],2)])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-0588fa57", __vue__options__)
  } else {
    hotAPI.reload("data-v-0588fa57", __vue__options__)
  }
})()}
