// smui/NumberBox/index.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".number-box {\n  display: inline-block;\n  position: relative;\n}\n.number-box input {\n  min-width: 4em;\n}\n.number-box input::-webkit-inner-spin-button {\n  -webkit-appearance: none;\n}\n.number-box .interator {\n  float: right;\n}\n.number-box .interator i {\n  position: absolute;\n  right: 0;\n  opacity: .3;\n  cursor: pointer;\n}\n.number-box .interator i:hover {\n  opacity: .8;\n}\n.number-box .interator .up {\n  transform: rotate(180deg);\n  top: 0;\n}\n.number-box .interator .down {\n  bottom: 0;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    value: '',
    min: 0,
    name: '',
    isInputing: true,
    max: window.Infinity,
    step: 0.1
};

exports.default = {
    params: params,
    extends: _ui2.default,
    methods: {
        stepUp: function stepUp() {
            this.$refs.ipt.stepUp();
        },
        stepDown: function stepDown() {
            this.$refs.ipt.stepDown();
        },
        changeValue: function changeValue($Events) {
            this.emit('model-change', $Events);
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"number-box"},[_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.value),expression:"value"}],ref:"ipt",attrs:{"type":"number","name":_vm.name,"step":_vm.step,"max":_vm.max,"min":_vm.min},domProps:{"value":_vm._s(_vm.value)},on:{"input":function($event){if($event.target.composing){ return; }_vm.value=_vm._n($event.target.value)},"blur":function($event){_vm.$forceUpdate()}}}),_vm._v(" "),_c('div',{staticClass:"interator"},[_c('i',{staticClass:"sm-arrow-dropdown up",on:{"click":_vm.stepUp}}),_vm._v(" "),_c('i',{staticClass:"sm-arrow-dropdown down",on:{"click":_vm.stepDown}})])])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-9a744ca6", __vue__options__)
  } else {
    hotAPI.reload("data-v-9a744ca6", __vue__options__)
  }
})()}
