// smui/RangeBar/Range.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".range-bar {\n  position: relative;\n}\n.range-bar .flag {\n  width: 2px;\n  height: 100%;\n  margin-left: -1px;\n  border-radius: 8px;\n  background-color: orange;\n  position: absolute;\n  top: 0;\n}\n.range-bar .flag span {\n  display: inline-block;\n  width: 100px;\n  position: absolute;\n  top: 1em;\n  left: -1em;\n}\n.range-bar input[type=range] {\n  width: 100%;\n}\n.range-bar .bar {\n  background-color: deepskyblue;\n  height: 10px;\n  border-radius: 5px;\n}\n.range-bar .range-box {\n  border: 1px solid #EEE;\n  border-radius: 5px;\n  background-color: #EEE;\n  width: 100%;\n  display: block;\n  position: relative;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    value: 0,
    height: 10,
    unit: '',
    showValue: false,
    type: 'custom',
    disabled: false,
    min: 0,
    max: 100
};

exports.default = {
    mixins: [_ui2.default],
    params: params,
    computed: {
        ratio: function ratio() {
            return Math.floor(this.value * 100 / (this.max - this.min)) + '%';
        }
    },
    methods: {
        changeValue: function changeValue(e) {
            if (this.disabled) {
                return false;
            }

            var el = this.$el.querySelector('.range-box');

            if (e && el) {
                this.value = (e.offsetX * this.max / el.clientWidth).toFixed(2);
            }
            this.emit('change-value', { value: this.value });
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"range-bar"},[(_vm.type == 'custom')?_c('div',{staticClass:"range-box",on:{"click":_vm.changeValue}},[(_vm.showValue)?_c('div',{staticClass:"min flag",style:({left: 0})},[_c('span',[_vm._v("0"+_vm._s(_vm.unit))])]):_vm._e(),_vm._v(" "),_c('div',{staticClass:"bar",style:({width: _vm.ratio, height: _vm.height+'px'})}),_vm._v(" "),(_vm.showValue)?_c('div',{staticClass:"flag",style:({left: _vm.ratio})},[_c('span',[_vm._v(_vm._s(_vm.value)+" "+_vm._s(_vm.unit))])]):_vm._e(),_vm._v(" "),(_vm.showValue)?_c('div',{staticClass:"max flag",style:({right: 0})},[_c('span',[_vm._v(_vm._s(_vm.max)+" "+_vm._s(_vm.unit))])]):_vm._e()]):_vm._e(),_vm._v(" "),_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.value),expression:"value"},{name:"show",rawName:"v-show",value:(_vm.type == 'native'),expression:"type == 'native'"}],attrs:{"type":"range","max":"max"},domProps:{"value":_vm._s(_vm.value)},on:{"change":_vm.changeValue,"input":function($event){_vm.value=$event.target.value}}})])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-ea5d42a8", __vue__options__)
  } else {
    hotAPI.reload("data-v-ea5d42a8", __vue__options__)
  }
})()}
