// smui/LineBox/LineBox.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".line-box-wrapper {\n  position: relative;\n  border: 1px solid #e4e4e4;\n  border-radius: 3px;\n  height: 302px;\n  overflow-y: scroll;\n  overflow-x: hidden;\n  background: #FFF;\n  font-size: 13px;\n}\n.line-box-wrapper .line-num {\n  display: block;\n  width: 30px;\n  text-align: center;\n  background-color: #e3e3e3;\n  position: absolute;\n  top: 0;\n  left: 0;\n  color: #666;\n}\n.line-box-wrapper .line-num .info-sign {\n  color: #C00;\n  cursor: pointer;\n}\n.line-box-wrapper .line-num li {\n  height: 28px;\n  line-height: 28px;\n  text-align: right;\n  padding-right: 8px;\n}\n.line-box-wrapper .info-sign .info.right {\n  min-height: 30px;\n  margin-top: 20px;\n  min-width: 150px;\n}\n.line-box-wrapper .info-sign .info.right:after,\n.line-box-wrapper .info-sign .info.right:before {\n  top: 7px;\n}\n.line-box-wrapper .input-textarea-container {\n  margin-left: 31px;\n}\n.line-box-wrapper .input-textarea[disabled] {\n  cursor: not-allowed;\n  background-color: #eee;\n}\n.line-box-wrapper .input-textarea {\n  width: 100%;\n  font-size: 13px;\n  border: 0;\n  outline: none;\n  line-height: 28px;\n  overflow: auto;\n  padding-left: 8px;\n  padding-top: 2px;\n  resize: none;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

var _Info = require('../Info');

var _Info2 = _interopRequireDefault(_Info);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    value: '',
    height: 300,
    width: 500,
    lineCount: 0,
    isInputing: true,
    errorInfo: {},
    disabled: false
};

exports.default = {
    private: {
        defaultText: '',
        textareaText: '' },
    params: params,
    mixins: [_ui2.default],
    ready: function ready() {
        this.errorInfo = {};
        this.value = this.defaultText || '';
    },

    watch: {
        textareaText: function textareaText(val) {
            this.value = val;
        },
        value: function value(val) {
            this.textareaText = val;
            this.textChanged();
        }
    },
    computed: {
        containerStyle: function containerStyle() {
            return {
                height: this.height + 'px',
                width: this.width + 'px'
            };
        },
        boxHeight: function boxHeight() {
            return Math.max(this.lineCount * 28 + 35, 400) + 'px';
        },
        txtStyle: function txtStyle() {
            return {
                height: this.boxHeight,
                width: this.width - 40 + 'px'
            };
        },
        textLine: function textLine() {
            return this.value.split('\n');
        }
    },
    methods: {
        setErrorObj: function setErrorObj(obj) {
            (0, _assign2.default)(this.errorInfo, obj);
        },
        textChanged: function textChanged(e) {
            var value = this.value;
            this.lineCount = value.split('\n').length;
            this.$emit('text-changed', value);
            this.emit('value-changed', { value: value });
        }
    },
    components: {
        Info: _Info2.default
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"line-box-wrapper",style:(_vm.containerStyle)},[_c('ul',{staticClass:"line-num",style:({height: _vm.boxHeight})},_vm._l((_vm.textLine),function(item,$index){return _c('li',[(_vm.errorInfo[item])?_c('info',{attrs:{"params":{direction:'right'}}},[_c('span',[_vm._v(_vm._s(_vm.errorInfo[item]))])]):_c('span',[_vm._v(_vm._s($index + 1))])],1)})),_vm._v(" "),_c('div',{staticClass:"input-textarea-container"},[_c('textarea',{directives:[{name:"model",rawName:"v-model",value:(_vm.value),expression:"value"}],staticClass:"input-textarea",style:(_vm.txtStyle),attrs:{"disabled":_vm.disabled,"wrap":"off"},domProps:{"value":_vm._s(_vm.value)},on:{"input":function($event){if($event.target.composing){ return; }_vm.value=$event.target.value}}})])])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-54d82ea7", __vue__options__)
  } else {
    hotAPI.reload("data-v-54d82ea7", __vue__options__)
  }
})()}
