// smui/Stepper/Stepper.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".stepper {\n  line-height: 25px;\n}\n.stepper li {\n  list-style: none;\n  display: inline-block;\n}\n.stepper .badget {\n  display: inline-block;\n  width: 25px;\n  height: 25px;\n  line-height: 25px;\n  text-align: center;\n  border-radius: 12.5px;\n  background-color: #CCC;\n  color: #FFF;\n}\n.stepper .badget:hover,\n.stepper .badget.active {\n  background-color: #26c899;\n}\n.stepper .text {\n  display: inline-block;\n}\n.stepper .line {\n  width: 100px;\n  border: 0;\n  height: 1px;\n  vertical-align: middle;\n  margin: 0 5px;\n  background-color: #CCC;\n  display: inline-block;\n}\n.stepper li:last-child .line {\n  display: none;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    value: 1,
    items: []
};

exports.default = {
    params: params,
    mixins: [_ui2.default],
    methods: {
        changeValue: function changeValue(value) {
            this.value = value;
            this.emit('value-change', { value: this.value });
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"stepper"},[_c('ul',_vm._l((_vm.items),function(item,index){return _c('li',[_c('div',{staticClass:"badget",class:{active: _vm.value > index },on:{"click":function($event){_vm.changeValue(index + 1)}}},[_vm._v(_vm._s(index + 1))]),_vm._v(" "),_c('div',{staticClass:"text"},[_vm._v(_vm._s(item.name))]),_vm._v(" "),_c('div',{staticClass:"line"})])}))])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-3f69c5db", __vue__options__)
  } else {
    hotAPI.reload("data-v-3f69c5db", __vue__options__)
  }
})()}
