// smui/DatePicker/config.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.cutList = exports.cutOption = exports.rangeMap = exports.defaultRange = exports.lastSeven = exports.lastOfMonth = exports.originDate = exports.firstDay = exports.lastQuater = exports.lastDay = exports.today = undefined;

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var format = 'YYYY-MM-DD';
var mday = (0, _moment2.default)();
var lastMonth = (0, _moment2.default)().add(-1, 'month');

/**
 * 时间变量
 */
var today = exports.today = mday.format(format);
var lastDay = exports.lastDay = (0, _moment2.default)().add(-1, 'days').format(format);
var lastQuater = exports.lastQuater = (0, _moment2.default)().add(-3, 'months').format(format);
var firstDay = exports.firstDay = (0, _moment2.default)().startOf('month').format(format);
var originDate = exports.originDate = (0, _moment2.default)().add(-18, 'month').format(format);

// 一个月第一天不能用昨天结束
var lastOfMonth = exports.lastOfMonth = lastDay;
if (mday.date() === 1) {
    exports.lastOfMonth = lastOfMonth = firstDay;
}

var lastSeven = exports.lastSeven = {
    begin: (0, _moment2.default)().subtract(1, 'week').format(format),
    end: (0, _moment2.default)().subtract(1, 'day').format(format)
};

var defaultRange = exports.defaultRange = [lastSeven.begin, lastSeven.end].join('~');

/**
 * 时间范围变量
 * @type {Object}
 */
var rangeMap = exports.rangeMap = {
    today: [today, today, 'today'],
    lastQuater: [lastQuater, today, 'lastQuater'],
    thisWeek: [mday.startOf('week').format(format), lastDay, 'thisWeek'],
    thisMonth: [firstDay, lastOfMonth, 'thisMonth'],
    lastDay: [lastDay, lastDay, 'lastDay'],
    lastWeek: [lastSeven.begin, lastSeven.end, 'lastWeek'],
    lastMonth: [lastMonth.startOf('month').format(format), lastMonth.endOf('month').format(format), 'lastMonth']
};

var cutOption = exports.cutOption = {
    lastQuater: 0,
    lastMonth: 1,
    thisMonth: 1,
    lastWeek: 1,
    thisWeek: 0,
    lastDay: 1,
    today: 1
};

var cutList = exports.cutList = [{
    name: 'lastQuater',
    text: 'lastQuater',
    value: rangeMap.lastQuater
}, {
    name: 'lastMonth',
    text: 'lastMonth',
    value: rangeMap.lastMonth
}, {
    name: 'thisMonth',
    text: 'thisMonth',
    value: rangeMap.thisMonth
}, {
    name: 'lastWeek',
    text: 'lastWeek',
    value: rangeMap.lastWeek
}, {
    name: 'thisWeek',
    text: 'thisWeek',
    value: rangeMap.thisWeek
}, {
    name: 'lastDay',
    text: 'lastDay',
    value: rangeMap.lastDay
}, {
    name: 'today',
    text: 'today',
    value: rangeMap.today
}];
