// smui/DatePicker/DatePicker.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".date-picker {\n  font-size: 12px;\n  font-weight: lighter;\n  position: relative;\n  color: #333;\n  line-height: 26px;\n  text-align: left;\n  height: 30px;\n  width: 205px !important;\n  display: inline-block;\n  border: 1px solid #CCC;\n  background-color: #FFF;\n}\n.date-picker * {\n  box-sizing: border-box;\n}\n.date-picker .button-panel {\n  position: absolute;\n  bottom: 10px;\n}\n.date-picker .info-sign {\n  position: absolute;\n  right: 45px;\n  top: 0px;\n}\n.date-picker .hidden {\n  display: none;\n}\n.date-picker .calable {\n  position: absolute;\n  display: block;\n  left: 0;\n  top: 0;\n  width: 100%;\n  height: 100%;\n  z-index: 3;\n}\n.date-picker .range-type {\n  text-align: left;\n}\n.date-picker .range-type a {\n  display: inline-block;\n  line-height: 15px;\n  padding: 0px 5px;\n  text-decoration: none;\n  cursor: pointer;\n}\n.date-picker .range-type a:hover,\n.date-picker .range-type .active {\n  color: #42b983;\n}\n.date-picker .layer {\n  position: absolute;\n  top: 31px;\n  left: 0;\n  background: #FFF;\n  width: 530px;\n  height: 345px;\n  padding: 10px;\n  border: 1px solid #EEE;\n  z-index: 100;\n}\n.date-picker .layer .button.button {\n  min-width: 60px;\n}\n.date-picker .date-wrapper {\n  position: relative;\n  display: inline-block;\n  width: 250px;\n}\n.date-picker .calendar {\n  left: 0;\n  top: 40px;\n  height: 220px;\n}\n.date-picker .range-input {\n  width: 100%;\n  padding-left: 5px;\n  text-align: left;\n}\n.date-picker .sm-calender {\n  position: absolute;\n  right: 16px;\n  top: 5px;\n  width: 5px;\n  height: 10px;\n  font-size: 15px;\n  color: #333;\n}\n.date-picker .date-input {\n  width: 238px;\n  line-height: 22px;\n  border: 1px solid #EEE;\n  border-radius: 2px;\n  background-color: #FFF;\n}\n.date-picker .close-btn {\n  position: absolute;\n  top: 10px;\n  right: 10px;\n  font-size: 12px;\n  font-weight: bolder;\n  cursor: pointer;\n}\n.date-picker .button {\n  min-width: 3em;\n}\n.date-picker .cancel-button {\n  color: #777 !important;\n  border: 1px solid #FFF;\n  background: #FFF;\n}\n.date-picker .calendar {\n  position: absolute;\n  border: 1px solid #DEDEDE;\n  border-radius: 2px;\n  transition: all .5s ease;\n}\n.date-picker .calendar:before {\n  position: absolute;\n  left: 30px;\n  top: -10px;\n  content: \"\";\n  border: 5px solid rgba(0, 0, 0, 0);\n  border-bottom-color: #DEDEDE;\n}\n.date-picker .calendar:after {\n  position: absolute;\n  left: 30px;\n  top: -9px;\n  content: \"\";\n  border: 5px solid rgba(0, 0, 0, 0);\n  border-bottom-color: #fff;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _Calendar = require('../Calendar');

var _Calendar2 = _interopRequireDefault(_Calendar);

var _Info = require('../Info');

var _Info2 = _interopRequireDefault(_Info);

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

var _i18n = require('./i18n');

var _config = require('./config');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    show: true,
    format: 'YYYY-MM-DD',
    info: '',
    limitDates: 365,
    value: '',
    rtype: '',
    syncModel: false,
    isInputing: true,
    cutList: _config.cutList,
    area: [_config.originDate, _config.today],
    areaRange: false,
    cutOption: _config.cutOption,
    cut: {}
};

exports.default = {
    params: params,
    lang: _i18n.lang,
    private: {
        showLayer: false,
        infoMap: {},
        begin: '',
        end: '',
        infoText: '' },
    mixins: [_ui2.default],
    data: function data() {
        var _defaultRange$split = _config.defaultRange.split('~'),
            _defaultRange$split2 = (0, _slicedToArray3.default)(_defaultRange$split, 2),
            begin = _defaultRange$split2[0],
            end = _defaultRange$split2[1];

        return {
            cutOption: (0, _assign2.default)({}, _config.cutOption),
            value: _config.defaultRange,
            begin: begin,
            end: end,
            cut: {}
        };
    },
    ready: function ready() {
        var _this = this;

        if (this.params) {
            (0, _assign2.default)(this, this.params);
        }

        this.cutList.forEach(function (item) {
            var key = item.value[0] + '~' + item.value[1];
            _this.infoMap[key] = item.name;
        });

        if (this.areaRange && this.areaRange.length > 2) {
            (function () {
                var unit = _this.areaRange[2] || 'days';
                _this.area = _this.areaRange.map(function (x) {
                    return (0, _moment2.default)().add(x, unit).format(_this.format);
                });
            })();
        }

        if (!this.area) {
            this.area = [_config.originDate, _config.today];
        }

        if (this.$el && this.$el.value) {
            this.value = this.$el.value;
        } else if (this.rtype) {
                var range = _config.rangeMap[this.rtype];
                this.setValue(range[0], range[1]);
            } else if (!this.value) {
                this.value = _config.defaultRange;
            }
        if (this.value) {
            this.updateInfo();
        }
    },

    watch: {
        model: function model(val) {
            if (val) {
                this.value = val;
            }
        },
        cut: function cut(val) {
            (0, _assign2.default)(this.cutOption, val);
        },
        value: function value(val) {
            this.updateInfo();
        }
    },
    computed: {
        beginParams: function beginParams() {
            return {
                value: this.begin,
                maxDate: this.end,
                minDate: false,
                area: this.beginArea,
                type: 'daterange',
                limitDates: this.limitDates,
                y: 36,
                show: this.show
            };
        },
        endParams: function endParams() {
            return {
                value: this.end,
                maxDate: false,
                minDate: this.begin,
                area: this.endArea,
                type: 'daterange',
                limitDates: this.limitDates,
                y: 36,
                show: this.show
            };
        },
        beginArea: function beginArea() {
            return this.area;
        },
        endArea: function endArea() {
            return this.area;
        }
    },
    methods: {
        ensure: function ensure() {
            this._oldValue = this.value;
            this.emit('range-change', {
                rtype: this.rtype,
                value: this.value,
                info: this.info,
                begin: this.begin,
                end: this.end
            });
            this.showLayer = false;
        },
        updateInfo: function updateInfo() {
            var _value$split = this.value.split('~'),
                _value$split2 = (0, _slicedToArray3.default)(_value$split, 2),
                begin = _value$split2[0],
                end = _value$split2[1];

            (0, _assign2.default)(this, { begin: begin, end: end });
            if (this.infoMap.hasOwnProperty(this.value)) {
                this.info = this.infoText = this.infoMap[this.value];
            } else {
                this.info = this.infoText = this.value;
            }
        },
        setRange: function setRange(value) {
            this.value = value;
            this.ensure();
        },
        setValue: function setValue(b, e) {
            this.value = (b || this.begin) + '~' + (e || this.end);
        },
        setBegin: function setBegin(e) {
            this.setValue(e.data.value);
        },
        setEnd: function setEnd(e) {
            this.setValue(0, e.data.value);
        },
        selectRange: function selectRange(item) {
            var _this2 = this;

            var range = item.value;
            this.setValue(range[0], range[1]);
            this.rtype = item.name;

            this.$nextTick(function () {
                _this2.info = _this2.infoText = item.name;
            });
        },
        select: function select(num, unit, text) {
            var m = (0, _moment2.default)();
            var end = '';
            if (unit === 'fmonth') {
                m = m.add(num, 'month');
                this.setValue(m.startOf('month').format(this.format), m.endOf('month').format(this.format));
            } else if (unit === 'day') {
                end = (0, _moment2.default)().add(num, 'days');
                this.setValue(end.format(this.format), end.format(this.format));
            } else {
                end = (0, _moment2.default)().add(-1, 'days');
                this.setValue(m.add(num, unit).format(this.format), end.format(this.format));
            }
            this.infoText = text;
        },
        showCalendar: function showCalendar(e) {
            e.stopPropagation();
            this.show = true;
            this._oldValue = this.value;
            this.showLayer = true;
        },
        clearTimer: function clearTimer() {
            clearTimeout(this.timer || 100);
        },
        hideCalendar: function hideCalendar(e) {
            var _this3 = this;

            if (e === 0) {
                this.showLayer = false;
                if (this._oldValue) {
                    this.value = this._oldValue;
                }
                return;
            }
            this.timer = setTimeout(function () {
                _this3.showLayer = false;
                if (_this3._oldValue) {
                    _this3.value = _this3._oldValue;
                }
            }, 400);
        }
    },
    components: { Calendar: _Calendar2.default, Info: _Info2.default }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"date-picker",on:{"mousemove":_vm.clearTimer,"mouseout":_vm.hideCalendar}},[_c('div',[_c('span',{staticClass:"range-input",attrs:{"id":"canlendar-button"},on:{"click":_vm.showCalendar}},[_vm._v(_vm._s(_vm.$lt(_vm.infoText)))]),_vm._v(" "),_c('label',{staticClass:"calable",attrs:{"for":"canlendar-button"},on:{"click":_vm.showCalendar}}),_vm._v(" "),_c('i',{staticClass:"sm-calender"})]),_vm._v(" "),_c('div',{staticClass:"layer",class:{hidden: !_vm.showLayer}},[_c('info',[_c('h3',[_vm._v(_vm._s(_vm.$lt("HELP_INFO")))]),_vm._v(" "),(_vm.limitDates)?_c('div',[_vm._v(_vm._s(_vm.$lt("LIMITED_DATES"))+" "+_vm._s(_vm.limitDates))]):_vm._e(),_vm._v(" "),(_vm.area && _vm.area.length > 1 && _vm.area[1])?_c('div',[_vm._v(_vm._s(_vm.$lt("RANGE_DATES"))+"\n                "),_c('div',[_vm._v(_vm._s(_vm.area[0])+" ~ "+_vm._s(_vm.area[1]))])]):_vm._e()]),_vm._v(" "),(_vm.cutList.length)?_c('div',{staticClass:"range-type"},_vm._l((_vm.cutList),function(c){return (_vm.cutOption[c.name])?_c('a',{class:{active: _vm.infoText == c.text},on:{"click":function($event){_vm.selectRange(c)}}},[_vm._v(_vm._s(_vm.$tr(c.name)))]):_vm._e()})):_vm._e(),_vm._v(" "),_c('div',[_c('div',{staticClass:"date-wrapper"},[_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.begin),expression:"begin"}],staticClass:"date-input start",attrs:{"type":"date","disabled":true},domProps:{"value":_vm._s(_vm.begin)},on:{"input":function($event){if($event.target.composing){ return; }_vm.begin=$event.target.value}}}),_vm._v(" "),_c('calendar',{attrs:{"params":_vm.beginParams},on:{"value-change":_vm.setBegin}})],1),_vm._v(" "),_c('div',{staticClass:"date-wrapper"},[_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.end),expression:"end"}],staticClass:"date-input end",attrs:{"type":"date","disabled":true},domProps:{"value":_vm._s(_vm.end)},on:{"input":function($event){if($event.target.composing){ return; }_vm.end=$event.target.value}}}),_vm._v(" "),_c('calendar',{attrs:{"params":_vm.endParams},on:{"value-change":_vm.setEnd}})],1)]),_vm._v(" "),_c('i',{staticClass:"close-btn close rotate sm-close",on:{"click":function($event){_vm.hideCalendar(0)}}}),_vm._v(" "),_c('div',{staticClass:"button-panel"},[_c('button',{staticClass:"ensure-button primary button",on:{"click":function($event){$event.preventDefault();_vm.ensure($event)}}},[_vm._v(_vm._s(_vm.$lt("ENSURE")))]),_vm._v(" "),_c('button',{staticClass:"cancel-button plain button",on:{"click":function($event){_vm.hideCalendar(0)}}},[_vm._v(_vm._s(_vm.$lt("CANCEL")))])])],1)])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-fdc27312", __vue__options__)
  } else {
    hotAPI.reload("data-v-fdc27312", __vue__options__)
  }
})()}
