// smui/DatePicker/doc.js
'use strict';

exports.info = {
    name: 'DatePicker',
    desc: '时间范围选择器',
    slot: false,
    inputing: true,
    detail: ['支持v-model绑定 <date-picker v-model="range"></date-picker>', '支持事件方式 <date-picker range-change="changeRange"></date-picker>', '支持选择范围设置-area', '支持时间跨度设置-limitDates', '支持自定义快捷时间范围选择-rtype/cut']
};

exports.events = [{ name: 'range-change', data: 'begin~end', desc: '日期范围发生变化' }];

exports.methods = {
    setValue: {
        args: 'begin, end',
        desc: '设置取值'
    },
    ensure: {
        args: '',
        desc: '确定选择'
    }
};

exports.params = {
    show: {
        default: false,
        type: ['Boolean'],
        desc: '是否显示浮层'
    },
    format: {
        default: 'YYYY-MM-DD',
        type: ['String'],
        desc: '日期格式'
    },
    limitDates: {
        default: '',
        type: ['Number'],
        desc: '最大时间跨度'
    },
    value: {
        default: '',
        type: ['String'],
        desc: '时间范围取值'
    },
    rtype: {
        default: '',
        type: ['String'],
        desc: '快捷日期选择，<br/>[lastQuater,lastMonth,thisMonth,lastWeek,thisWeek,lastDay,today]'
    },
    syncModel: {
        default: false,
        type: ['Boolean'],
        desc: '是否自动同步；默认需要点击确定后同步值'
    },
    cutList: {
        default: 'config.cutList',
        type: ['Array'],
        desc: '可选择的快捷操作列表'
    },
    area: {
        default: ['一年半前', '今天'],
        type: ['Array'],
        desc: '可选择时间范围'
    },
    areaRange: {
        default: false,
        type: ['Array'],
        desc: '可选择时间范围计算 [-3, 0, "months"]'
    },
    cut: {
        default: 'config.cutOption',
        type: ['Object'],
        desc: '便捷操作显示配置 <br/>{"lastQuater":0,"lastMonth":1,"thisMonth":1,"lastWeek":1,"thisWeek":0,"lastDay":1,"today":1}'
    }
};
