// smui/SelectList/SelectList.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".select-wrapper {\n  position: relative;\n  display: inline-block;\n  box-sizing: border-box;\n  cursor: pointer;\n  line-height: 30px;\n  background: rgba(255, 255, 255, 0.5);\n}\n.select-wrapper .select-tip {\n  position: absolute;\n  top: 0;\n  left: 0;\n  line-height: 30px;\n  padding: 0 10px;\n  color: #333;\n}\n.select-wrapper .drop-icon {\n  pointer-events: none;\n  position: absolute;\n  right: 8px;\n  top: 10px;\n  font-size: 12px;\n  color: #333;\n  z-index: 7;\n}\n.select-wrapper .select {\n  padding: 0 10px;\n  padding-right: 40px;\n  background-color: transparent;\n  box-sizing: border-box;\n  border: 1px solid #c0c8c9;\n  border-radius: 4px;\n  position: relative;\n  background-repeat: no-repeat;\n  background-size: 1.5rem 1rem;\n  background-position: right center;\n  font: inherit;\n  color: inherit;\n  z-index: 3;\n  height: 30px;\n  -webkit-appearance: none;\n  -moz-appearance: none;\n  appearance: none;\n}\n.select-wrapper .select:focus {\n  outline: none;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    name: '',
    items: [],
    isInputing: true,
    disabled: false,
    title: '',
    value: ''
};

exports.default = {
    mixins: [_ui2.default],
    params: params,
    data: function data() {
        var uid = (Math.random().toFixed(3) * 1000).toString(32);
        return {
            uiId: 'select_' + uid
        };
    },

    watch: {
        'params.items': function paramsItems(value) {
            this.items = value;
        }
    },
    mounted: function mounted() {
        if (this.value === '' && this.items.length) {
            this.value = this.items[0].value;
            this.name = this.name || this.items[0].name;
        }
    },

    methods: {
        getValue: function getValue() {
            return this.value;
        },
        changeValue: function changeValue(e) {
            var select = this.$el.querySelector('select');
            var idx = 0;
            var name = '';
            if (select) {
                idx = select.options.selectedIndex;
                name = (select.options[idx] || {}).textContent || '';
            }
            this.emit('value-change', {
                value: this.value,
                index: idx,
                name: name
            });
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"select-wrapper"},[(_vm.name && !_vm.value)?_c('div',{staticClass:"select-tip"},[_vm._v(_vm._s(_vm.name))]):_vm._e(),_vm._v(" "),_c('select',{directives:[{name:"model",rawName:"v-model",value:(_vm.value),expression:"value"}],staticClass:"select",attrs:{"id":_vm.uiId,"disabled":_vm.disabled},on:{"change":[function($event){_vm.value=Array.prototype.filter.call($event.target.options,function(o){return o.selected}).map(function(o){var val = "_value" in o ? o._value : o.value;return val})[0]},_vm.changeValue]}},[_vm._t("default",[(_vm.title)?_c('option',{attrs:{"disabled":"","value":""},domProps:{"value":""}},[_vm._v(_vm._s(_vm.title))]):_vm._e(),_vm._v(" "),_vm._l((_vm.items),function(item){return _c('option',{domProps:{"value":item.value}},[_vm._v(_vm._s(item.name))])})])],2),_vm._v(" "),_c('i',{staticClass:"drop-icon sm-arrow-dropdown",attrs:{"for":_vm.uiId}})])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-7c90efb7", __vue__options__)
  } else {
    hotAPI.reload("data-v-7c90efb7", __vue__options__)
  }
})()}
