// smui/PressTab/PressTab.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".press-tab {\n  height: 30px;\n  display: inline-block;\n  border: 1px solid #CCC;\n  border-radius: 5px;\n  background-color: #FFF;\n  margin: 10px 18px;\n}\n.press-tab a {\n  display: inline-block;\n  width: 84px;\n  line-height: 28px;\n  border-right: 1px solid #CCC;\n  padding: 0 5px;\n  text-align: center;\n  background-color: #FFF;\n  float: left;\n  color: #333;\n  background: linear-gradient(#FFF, #ebebeb);\n  cursor: pointer;\n}\n.press-tab a.active {\n  background: linear-gradient(#ebebeb, #FFF);\n  cursor: auto;\n}\n.press-tab a:first-child {\n  border-radius: 5px 0px 0px 5px;\n  border-right: 1px solid #CCC;\n}\n.press-tab a:last-child {\n  border-radius: 0px 5px 5px 0;\n  border-right: 0;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    items: [],
    isInputing: true,
    tab: ''
};

exports.default = {
    params: params,
    mixins: [_ui2.default],
    methods: {
        changeTab: function changeTab(tab) {
            this.tab = tab;
            this.emit('tab-change', { tab: tab });
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"press-tab"},[_vm._t("default",_vm._l((_vm.items),function(item){return _c('a',{class:{active: _vm.tab == item.name},attrs:{"value":item.value},on:{"click":function($event){_vm.changeTab(item.name)}}},[_vm._v(_vm._s(item.text || item.name))])}))],2)}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-629f03f7", __vue__options__)
  } else {
    hotAPI.reload("data-v-629f03f7", __vue__options__)
  }
})()}
