// smui/Schedule/Schedule.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".select-timer-control {\n  font-size: 13px;\n  width: 850px;\n  background: #fff;\n  padding: 10px;\n  margin: 0 auto;\n}\n.select-timer-control .tip-bar {\n  height: 18px;\n  line-height: 18px;\n  padding: 10px 0;\n}\n.select-timer-control .tip-bar .action {\n  display: inline-block;\n}\n.select-timer-control .tip-bar .action span {\n  display: inline-block;\n  padding: 0 5px;\n  cursor: pointer;\n}\n.select-timer-control .tip-bar .tip {\n  display: inline-block;\n  float: right;\n  margin-left: 20px;\n}\n.select-timer-control .tip-bar .tip span {\n  display: inline-block;\n  height: 13px;\n  margin-top: 2px;\n  width: 13px;\n  vertical-align: top;\n  margin-right: 5px;\n}\n.select-timer-control .tip-bar .tip .selected-tip {\n  background-color: #26bc89;\n}\n.select-timer-control .tip-bar .tip .unselected-tip {\n  background-color: #ccc;\n}\n.select-timer-control .week-container {\n  font-size: 0;\n}\n.select-timer-control .week-container .day-line {\n  display: block;\n}\n.select-timer-control .week-container .day-th-line {\n  margin-top: 16px;\n}\n.select-timer-control .week-container .day-container li {\n  display: inline-block;\n}\n.select-timer-control .week-container .day-container .week-th {\n  width: 88px;\n  font-size: 13px;\n  box-sizing: border-box;\n  height: 30px;\n  line-height: 30px;\n}\n.select-timer-control .week-container .day-container .week-th label {\n  cursor: pointer;\n}\n.select-timer-control .week-container .day-container .hour-th,\n.select-timer-control .week-container .day-container .hour-td {\n  width: 30px;\n  height: 30px;\n  vertical-align: bottom;\n  overflow: hidden;\n  border: 1px solid #fff;\n  border-width: 1px 1px 0 0;\n  box-sizing: border-box;\n}\n.select-timer-control .week-container .day-container .hour-td {\n  color: #fff;\n  text-align: center;\n  line-height: 30px;\n  font-size: 12px;\n  cursor: pointer;\n}\n.select-timer-control .week-container .day-container .selected-0 {\n  background-color: #ccc;\n}\n.select-timer-control .week-container .day-container .selected-1 {\n  background-color: #26bc89;\n}\n.select-timer-control .week-container .day-container .selected-1:hover {\n  background-color: #fbd4ba;\n}\n.select-timer-control .week-container .day-container .hour-th {\n  font-size: 14px;\n  text-align: center;\n  line-height: 30px;\n  cursor: pointer;\n  color: #ccc;\n}\n.select-timer-control .week-container .day-container .hour-th:hover,\n.select-timer-control .week-container .day-container .hour-th.selected {\n  color: #26bc89;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

var _i18n = require('./i18n');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var pd = 'd';

var ph = 'h';

var map = function () {
    var map = {};
    for (var i = 0; i < 7; i++) {
        map[pd + i] = {};
        for (var j = 0; j < 24; j++) {
            map[pd + i][ph + j] = 0;
        }
    }
    return map;
}();
var params = {};

var toolbarCheckTimer = null;

exports.default = {
    mixins: [_ui2.default],
    lang: _i18n.lang,
    params: params,
    data: function data() {
        var hLine = {};
        var sLine = {};
        for (var i = 0; i < 7; i++) {
            for (var j = 0; j < 24; j++) {
                hLine['h' + i] = 0;
                sLine['s' + j] = 0;
            }
        }
        var data = {
            actionFlags: {
                sLine: sLine,
                hLine: hLine
            },
            defaultPeriod: '',
            valueMap: map,
            weekName: ['MON', 'TUE', 'WEN', 'THU', 'FRI', 'SAT', 'SUN']
        };
        return data;
    },

    methods: {
        getSelectedString: function getSelectedString() {
            var map = this.valueMap;
            var valueArr = [];
            var startStr = '';
            var endStr = '';
            function joinStr(day, hour, value) {
                if (value) {
                    var vStr = (day + 1) * 100 + hour;
                    !startStr && (startStr = vStr);
                    endStr = vStr + 1;
                }

                if ((!value || hour === 23) && startStr) {
                    valueArr.push(startStr + '-' + endStr);
                    startStr = endStr = '';
                }
            };

            for (var i = 0; i < 7; i++) {
                for (var j = 0; j < 24; j++) {
                    joinStr(i, j, map[pd + i][ph + j]);
                }
            }
            return valueArr.join(';');
        },
        checkeToolBarStatus: function checkeToolBarStatus() {
            var me = this;
            function doCheck() {
                var map = me.valueMap;
                var af = me.actionFlags;

                var recodeObj = {};
                for (var i = 0; i < 7; i++) {
                    for (var j = 0; j < 24; j++) {
                        var tvalue = map[pd + i][ph + j];
                        if (!recodeObj['h' + i]) {
                            af.hLine['h' + i] = tvalue;
                            !tvalue && (recodeObj['h' + i] = 1);
                        };
                        if (!recodeObj['s' + j]) {
                            af.sLine['s' + j] = tvalue;
                            !tvalue && (recodeObj['s' + j] = 1);
                        };
                    }
                }
            }
            clearTimeout(toolbarCheckTimer);
            toolbarCheckTimer = setTimeout(doCheck, 200);
        },
        setValueMap: function setValueMap(valueObj) {
            var me = this;
            var map = me.valueMap;
            var af = me.actionFlags;

            for (var i = 0; i < 7; i++) {
                valueObj[pd + i] = valueObj[pd + i] || {};
                for (var j = 0; j < 24; j++) {
                    var tvalue = valueObj[pd + i][ph + j];
                    tvalue = tvalue ? tvalue : 0;
                    map[pd + i][ph + j] = tvalue;

                    if (!(af.hLine['h' + i] === 0)) {
                        af.hLine['h' + i] = tvalue;
                    };
                    if (!(af.sLine['s' + j] === 0)) {
                        af.hLine['s' + j] = tvalue;
                    };
                }
            }
            me.checkeToolBarStatus();
        },

        setDefaultPeriod: function setDefaultPeriod(period) {
            if (period && typeof period === 'string') {
                this.defaultPeriod = period;
                this.resetValue();
            }
        },
        resetValue: function resetValue() {
            var valueObj = this.setStrToValueObj(this.defaultPeriod);
            this.setValueMap(valueObj);
        },
        setActionArea: function setActionArea(flag) {
            var tagObj = {
                workday: 1,
                weekend: 0,
                all: 2
            };
            var tag = +tagObj[flag];
            for (var i = 0; i < 7; i++) {
                this.setDayLine(i, i < 5 && tag || i >= 5 && !tag || tag == 2);
            }
        },
        setDayLine: function setDayLine(day, value) {
            var me = this;
            var tvalue = +!me.actionFlags.hLine['h' + day];
            if ('undefined' != typeof value) {
                tvalue = +!!value;
            };
            for (var i = 0; i < 24; i++) {
                me.valueMap[pd + day][ph + i] = tvalue;
            }
            me.checkeToolBarStatus();
        },
        setHourLine: function setHourLine(hour) {
            var me = this;
            var tvalue = +!me.actionFlags.sLine['s' + hour];
            for (var i = 0; i < 7; i++) {
                me.valueMap[pd + i][ph + hour] = tvalue;
            }
            me.checkeToolBarStatus();
        },
        setSingleActionFlag: function setSingleActionFlag(day, hour) {
            var me = this;
            var map = me.valueMap;
            var svalue = 1,
                hvalue = 1;
            if (!map[pd + day][ph + hour]) {
                svalue = 0;
                hvalue = 0;
            } else {
                for (var i = 0; i < 7; i++) {
                    if (!map[pd + i][ph + hour]) {
                        svalue = 0;
                        break;
                    };
                }
                for (i = 0; i < 24; i++) {
                    if (!map[pd + day][ph + i]) {
                        hvalue = 0;
                        break;
                    }
                }
            }
            me.actionFlags.sLine['s' + hour] = svalue;
            me.actionFlags.hLine['h' + day] = hvalue;
        },
        setSingleHour: function setSingleHour(day, hour) {
            var cValue = this.valueMap[pd + day][ph + hour];
            this.valueMap[pd + day][ph + hour] = +!cValue;
            this.setSingleActionFlag(day, hour);
        },
        setStrToValueObj: function setStrToValueObj(valueStr) {
            var valueObj = {};
            if (valueStr) {
                valueStr.split(';').map(function (item) {
                    var cstr = item.split('-');
                    var d = item[0] - 1;
                    valueObj[pd + d] = valueObj[pd + d] || {};
                    var start = +cstr[0].substr(1, 2);
                    var end = +cstr[1].substr(1, 2);
                    for (var i = start, j = end; i < j; i++) {
                        valueObj[pd + d][ph + i] = 1;
                    }
                });
            };
            return valueObj;
        }
    },
    compiled: function compiled() {
        this.resetValue();
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"select-timer-control"},[_c('div',{staticClass:"tip-bar"},[_c('p',{staticClass:"action"},[_c('span',{on:{"click":function($event){_vm.setActionArea('workday')}}},[_vm._v(_vm._s(_vm.$lt('WORK_DAY')))]),_vm._v("|\n            "),_c('span',{on:{"click":function($event){_vm.setActionArea('weekend')}}},[_vm._v(_vm._s(_vm.$lt('WEEK_END')))]),_vm._v("|\n            "),_c('span',{on:{"click":function($event){_vm.setActionArea('all')}}},[_vm._v(_vm._s(_vm.$lt('ALL')))]),_vm._v("|\n            "),_c('span',{on:{"click":function($event){_vm.resetValue()}}},[_vm._v(_vm._s(_vm.$lt('RESET')))])]),_vm._v(" "),_c('p',{staticClass:"tip"},[_c('span',{staticClass:"unselected-tip"}),_vm._v(_vm._s(_vm.$lt('STOP_RANGE')))]),_vm._v(" "),_c('p',{staticClass:"tip"},[_c('span',{staticClass:"selected-tip"}),_vm._v(_vm._s(_vm.$lt('ACTIVE_RANGE')))])]),_vm._v(" "),_c('div',[_c('ul',{staticClass:"week-container"},[_c('li',{staticClass:"day-line day-th-line"},[_c('ul',{staticClass:"day-container"},[_c('li',{staticClass:"week-th"}),_vm._v(" "),_vm._l((24),function(hour){return _c('li',{class:{'hour-th':1,'selected':_vm.actionFlags.sLine['s'+(hour-1)]},on:{"click":function($event){_vm.setHourLine((hour-1))}}},[_c('span',{staticClass:"sm-download"})])})],2)]),_vm._v(" "),_vm._l((7),function(day,$index){return _c('li',{staticClass:"day-line"},[_c('ul',{staticClass:"day-container"},[_c('li',{staticClass:"week-th",attrs:{"data-value":$index}},[_c('label',[_c('input',{attrs:{"type":"checkbox","name":"week-day"},domProps:{"checked":!!_vm.actionFlags.hLine['h'+(day-1)]},on:{"click":function($event){_vm.setDayLine((day-1))}}}),_vm._v("\n                            "+_vm._s(_vm.$lt(_vm.weekName[(day-1)]))+"\n                        ")])]),_vm._v(" "),_vm._l((24),function(hour){return _c('li',{class:['hour-td', 'selected-'+_vm.valueMap['d'+(day-1)]['h'+(hour-1)]],on:{"click":function($event){_vm.setSingleHour((day-1), (hour-1))}}},[_vm._v("\n                        "+_vm._s(hour-1)+"\n                    ")])})],2)])})],2)])])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-680bf197", __vue__options__)
  } else {
    hotAPI.reload("data-v-680bf197", __vue__options__)
  }
})()}
