// smui/Dialog/Dialog.js
'use strict';

var _util = require('../base/util');

var _util2 = _interopRequireDefault(_util);

var _index = require('../DialogPanel/index');

var _index2 = _interopRequireDefault(_index);

var _Prompt = require('./Prompt');

var _Prompt2 = _interopRequireDefault(_Prompt);

var _Toast = require('./Toast');

var _Toast2 = _interopRequireDefault(_Toast);

var _Loading = require('./Loading');

var _Loading2 = _interopRequireDefault(_Loading);

var _Popup = require('./Popup');

var _Popup2 = _interopRequireDefault(_Popup);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Dialog = _util2.default.wrapper(_index2.default);
Dialog.DialogPanel = _index2.default;
Dialog.Toast = _Toast2.default;

window.Dialog = Dialog;

Dialog.params = _index2.default.params;

Dialog.toast = function (message, option) {
    if (typeof message === 'string') {
        option = Object.assign({
            message: message
        }, option || {});
    } else if (message) {
        option = Object.assign(message, option || {});
    }
    var Factor = _util2.default.wrapper(_Toast2.default);
    var toast = new Factor();
    toast.show(option);
    return toast;
};

Dialog.info = function (message, option) {
    Dialog.toast({ message: message, clazz: 'info' }, option);
};

Dialog.error = function (message, option) {
    Dialog.toast({ message: message, clazz: 'error' }, option);
};

Dialog.warn = function (message, option) {
    Dialog.toast({ message: message, clazz: 'warn' }, option);
};

Dialog.popup = function (option) {
    var Factor = _util2.default.wrapper(_Popup2.default);
    var pop = new Factor();
    pop.show(option);
    return pop;
};

Dialog.loading = function (option) {
    var Factor = _util2.default.wrapper(_Loading2.default);
    var loading = new Factor();
    loading.show(option);
    return loading;
};

Dialog.alert = function (message, ensure) {
    var width = 250;
    if (typeof message === 'string') {
        if (message.length > 20) {
            width = 400;
        }
        message = {
            content: message,
            width: width,
            clazz: ['smui', 'alert'],
            needClose: false,
            needCancel: false
        };
    }

    var dialog = new Dialog({
        data: Object.assign({
            width: width,
            needClose: false,
            needCancel: false,
            clazz: ['smui', 'alert']
        }, message)
    });

    dialog.showModal(message);
    dialog.$on('ensure', function () {
        ensure && ensure.call(dialog);
        dialog.close();
    });
    return dialog;
};

Dialog.confirm = function (message, ensure, cancel) {
    var width = 250;
    if (typeof message === 'string') {
        if (message.length > 20) {
            width = 400;
        }
        message = {
            content: message
        };
    }

    var dialog = new Dialog({
        data: Object.assign({
            width: width,
            needClose: false,
            needCancel: true,
            clazz: ['smui', 'confirm']
        }, message)
    });

    dialog.showModal(message);
    dialog.$on('ensure', function () {
        ensure && ensure.call(dialog);
        dialog.close({ isMute: true });
    });

    dialog.$on('closed', function () {
        cancel && cancel.call(dialog);
    });
    return dialog;
};

Dialog.prompt = function (option, ensure, cancel) {
    if (typeof option === 'string') {
        option = {
            title: option,
            type: 'text'
        };
    }

    if (option.options) {
        option.type = 'select';
    }
    if (!option.type) {
        option.type = 'text';
    }
    if (!option.error) {
        option.error = '';
    }
    option.needClose = false;

    var c = option.container;

    var dialog = new Dialog({
        container: c,
        data: option,
        content: _Prompt2.default
    });

    if (option.container) {
        dialog.show(option);
    } else {
        dialog.showModal();
    }

    dialog.$on('finished', function (e) {
        ensure && ensure.call(dialog, e.value);
        dialog.close();
    });

    dialog.$on('cancel', function () {
        cancel && cancel.call(dialog);
        dialog.close();
    });
    return dialog;
};

Dialog.inlineEdit = function (option, ensure, cancel) {
    option = Object.assign(option, { inline: true, needClose: true, needFooter: true });
    return Dialog.prompt(option, ensure, cancel);
};

module.exports = Dialog;
