// smui/Dialog/doc.js
'use strict';

exports.info = {
    name: 'Dialog',
    desc: '弹窗组件',
    detail: ['指令创建 new Dialog().show()', '标签创建 <dialog-panel>...</dialog-panel>', '一般展现: show()', '模态展现: showModal()']
};

exports.slots = [{ name: 'title', desc: '标题' }, { name: 'content', desc: '内容' }, { name: 'footer', desc: '脚底' }];

exports.events = [{ name: 'ensure', data: '*', desc: '确认操作' }, { name: 'closed', data: '*', desc: '关闭完成' }, { name: 'closing', data: '*', desc: '关闭前' }, { name: '*', data: '*', desc: '子组件emit事件' }];

exports.methods = {
    show: {
        args: '{options}',
        desc: '普通展现'
    },
    showModal: {
        args: '{options}',
        desc: '模态展现'
    },
    close: {
        args: '',
        desc: '关闭弹窗'
    },
    hide: {
        args: '',
        desc: '隐藏弹窗'
    }
};

exports.params = {
    title: {
        default: '',
        type: ['String'],
        desc: '标题'
    },
    content: {
        default: '',
        type: ['String'],
        desc: '内容'
    },
    ensureText: {
        default: '确认',
        type: ['String'],
        desc: '确认文本'
    },
    cancelText: {
        default: '取消',
        type: ['String'],
        desc: '取消文本'
    },
    data: {
        default: '子组件传入',
        type: ['String'],
        desc: '子组件需要props.params参数'
    },
    clazz: {
        default: '{}',
        type: ['Object', 'String', 'Array'],
        desc: '追加的Class名称，为了方便扩展样式'
    },
    style: {
        default: '{}',
        type: ['Object'],
        desc: '容器对象追加样式'
    },
    width: {
        default: 'auto',
        type: ['Number'],
        desc: '宽度'
    },
    height: {
        default: 'auto',
        type: ['Number'],
        desc: '高度'
    },
    headerView: {
        default: 'head',
        type: ['String'],
        desc: '自定义Header视图'
    },
    contentView: {
        default: 'ContentView',
        type: ['String'],
        desc: '自定义Content视图'
    },
    footerView: {
        default: 'foot',
        type: ['String'],
        desc: '自定义Footer视图'
    },
    removeOnClose: {
        default: 'true',
        type: ['String'],
        desc: '关闭时是否删除元素'
    },
    needClose: {
        default: 'true',
        type: ['Boolean'],
        desc: '是否需要X关闭'
    },
    needFooter: {
        default: 'true',
        type: ['Boolean'],
        desc: '是否需要Footer'
    },
    needCancel: {
        default: 'true',
        type: ['Boolean'],
        desc: '是否需要Cancel'
    }
};
