// smui/Dialog/Loading.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".loading-box.mask {\n  width: 100%;\n  height: 100%;\n  background: rgba(0, 0, 0, 0.2);\n}\n.loading-box.full {\n  position: absolute;\n  width: 100%;\n  height: 100%;\n  position: fixed;\n}\n.loading-box.full .content {\n  position: absolute;\n  top: 50%;\n  left: 49vw;\n}\n.icon-snake {\n  border: 4px solid transparent;\n  border-top-color: #26a2ff;\n  border-left-color: #26a2ff;\n  border-bottom-color: #26a2ff;\n  height: 28px;\n  width: 28px;\n  display: inline-block;\n  border-radius: 50%;\n}\n.loading-box {\n  z-index: 10001;\n  top: 0;\n  right: 0;\n  bottom: 0;\n  left: 0;\n  position: relative;\n  min-height: 5em;\n  vertical-align: top;\n  background: rgba(0, 0, 0, 0.01);\n}\n.loading-box dialog {\n  border: 0;\n  background: rgba(255, 255, 255, 0.01);\n  top: 50%;\n  transform: translateY(-50%);\n}\n.loading-box .close {\n  font-size: 1.5em;\n}\n.loading-box i.sm-icon {\n  font-size: 3em;\n  color: #26a2ff;\n  margin-left: -20px;\n  margin-top: -20px;\n}\n.loading-box div.sm-icon {\n  width: 100%;\n  left: 0;\n  text-align: center;\n}\n.loading-box .sm-icon {\n  position: absolute;\n  left: 50%;\n  top: 50%;\n  animation: loading 1s infinite step-start;\n}\n.loading-box .content {\n  text-align: center;\n  margin-top: 6em;\n}\n.loading-box .close {\n  position: absolute;\n  top: 5px;\n  right: 5px;\n}\n@keyframes loading {\n  0% {\n    transform: rotate(0deg);\n  }\n  12.5% {\n    transform: rotate(45deg);\n  }\n  25% {\n    transform: rotate(90deg);\n  }\n  37.5% {\n    transform: rotate(135deg);\n  }\n  50% {\n    transform: rotate(180deg);\n  }\n  62.5% {\n    transform: rotate(225deg);\n  }\n  75% {\n    transform: rotate(270deg);\n  }\n  87.5% {\n    transform: rotate(315deg);\n  }\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    mixins: [_ui2.default],
    params: {
        isShow: false,
        isFull: true,
        needMask: true,
        showMask: false,
        content: '',
        timeout: 10000,
        delay: 200,
        icon: 'sm-loading'
    },
    methods: {
        show: function show(option) {
            var _this = this;

            if (option) {
                (0, _assign2.default)(this, option);
            }
            this.isShow = true;
            setTimeout(function () {
                _this.close();
            }, this.timeout);

            var $vm = this;
            this.isShow = true;

            if (this.needMask) {
                setTimeout(function () {
                    $vm.showMask = true;
                    var dialog = $vm.$el.querySelector('dialog');
                    if (dialog.isConnected && dialog) {
                        if ($vm.isFull) {
                            dialog.open = false;
                            try {
                                dialog.showModal();
                            } catch (ex) {
                                dialog.open = true;
                            }
                        } else {
                            dialog.show();
                        }
                    }
                }, this.delay);
            }
        },
        close: function close() {
            if (this.$el) {
                this.$el.remove();
            }
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{directives:[{name:"show",rawName:"v-show",value:(_vm.isShow),expression:"isShow"}],staticClass:"loading-box",class:{mask: _vm.showMask, full: _vm.isFull}},[_c('dialog',{attrs:{"open":""}},[(_vm.icon)?_c('i',{staticClass:"sm-icon",class:_vm.icon}):_vm._e(),_vm._v(" "),_c('div',{staticClass:"content"},[_vm._t("default",[_vm._v(_vm._s(_vm.content))])],2)])])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-4b8cfe1d", __vue__options__)
  } else {
    hotAPI.reload("data-v-4b8cfe1d", __vue__options__)
  }
})()}
