// smui/Dialog/Popup.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".popup {\n  position: fixed;\n  z-index: 9999;\n  bottom: 10px;\n  right: 10px;\n  font-size: 12px;\n  border: 1px solid #CCC;\n  border-radius: 5px;\n  min-width: 250px;\n  background: #FFF;\n}\n.popup .close {\n  position: absolute;\n  top: 5px;\n  right: 5px;\n}\n.popup .title {\n  border-bottom: 1px solid #CCC;\n  line-height: 2em;\n  padding: 5px 10px;\n}\n.popup .content {\n  padding: 40px 10px 20px;\n  min-height: 30px;\n  max-width: 300px;\n}\n.popup .footer {\n  height: 40px;\n}\n.popup .close-button {\n  float: right;\n  color: #FFF;\n  border: 1px solid #39bd8b;\n  background: linear-gradient(rgba(57, 189, 139, 0.7), #39bd8b);\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

var _layer = require('../base/layer');

var _layer2 = _interopRequireDefault(_layer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var counter = 0;
exports.default = {
    mixins: [_ui2.default, _layer2.default],
    params: {
        message: '',
        title: '',
        delay: 2000,
        needClose: false,
        index: 0,
        bias: [0, 10],
        isShow: false
    },
    computed: {
        style: function style() {
            var index = this.index;

            var _bias = (0, _slicedToArray3.default)(this.bias, 2),
                bl = _bias[0],
                bt = _bias[1];

            return {
                bottom: index * bt + 'px',
                right: index * bl + 'px'
            };
        }
    },
    methods: {
        show: function show(option) {
            (0, _assign2.default)(this, option);
            this.isShow = true;
            this.index = counter++;
        },
        close: function close() {
            this.isShow = false;
            counter--;
            this.emit('close');
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return (_vm.isShow)?_c('div',{staticClass:"popup",class:_vm.clazz,style:(_vm.style)},[_c('i',{staticClass:"sm-close close",on:{"click":_vm.close}}),_vm._v(" "),(_vm.title)?_c('div',{staticClass:"title"},[_vm._v(_vm._s(_vm.title))]):_vm._e(),_vm._v(" "),_c('div',{staticClass:"content",domProps:{"innerHTML":_vm._s(_vm.message)}}),_vm._v(" "),(_vm.needClose)?_c('div',{staticClass:"footer"},[_c('div',{staticClass:"close-button button",on:{"click":_vm.close}},[_vm._v(_vm._s(_vm.$lt('CLOSE')))])]):_vm._e()]):_vm._e()}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-6904d3cd", __vue__options__)
  } else {
    hotAPI.reload("data-v-6904d3cd", __vue__options__)
  }
})()}
