// smui/Dialog/Toast.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert("@keyframes toast-opacity {\n  0% {\n    opacity: 0;\n  }\n  50% {\n    opacity: 1;\n  }\n  75% {\n    opacity: 0.75;\n  }\n  100% {\n    opacity: 0;\n  }\n}\n.dialog-panel.smui.alert,\n.dialog-panel.smui.confirm {\n  line-height: 2em;\n}\n.toast {\n  position: fixed;\n  top: 50%;\n  left: 50%;\n  z-index: 99999;\n  transform: translateX(-50%) translateY(-50%);\n  text-align: center;\n  font-size: 12px;\n  padding: 5px 10px;\n  background: #EEE;\n  border: 1px solid #CCC;\n  border-radius: 5px;\n  animation: toast-opacity 2s;\n}\n.toast.info {\n  border: 1px solid #d5f1fd;\n  background-color: #eaf8fe;\n}\n.toast.tip {\n  border: 1px solid #e7f6e1;\n  background-color: #f3faf0;\n  color: #777;\n}\n.toast.warn {\n  border: 1px solid #fec;\n  background-color: #fff7e6;\n  color: #d00;\n}\n.toast.error {\n  border: 1px solid #fdc;\n  background-color: #ffeee6;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

var _layer = require('../base/layer');

var _layer2 = _interopRequireDefault(_layer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    mixins: [_ui2.default, _layer2.default],
    params: {
        message: '',
        title: '',
        delay: 2000,
        isShow: true,
        clazz: ['info']
    },
    methods: {
        show: function show(option) {
            var _this = this;

            this.isShow = true;
            (0, _assign2.default)(this, option);
            setTimeout(function () {
                _this.isShow = false;
            }, this.delay);
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{directives:[{name:"show",rawName:"v-show",value:(_vm.isShow),expression:"isShow"}],staticClass:"toast",class:_vm.clazz,style:(_vm.style)},[_vm._t("default",[_c('div',[_vm._v(_vm._s(_vm.title))]),_vm._v(" "),_c('div',[_vm._v(_vm._s(_vm.message))])])],2)}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-c5307870", __vue__options__)
  } else {
    hotAPI.reload("data-v-c5307870", __vue__options__)
  }
})()}
