// @gmod/bgzf-filehandle/dist/unzip.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.pakoUnzip = exports.nodeUnzip = exports.unzipChunkSlice = exports.unzipChunk = exports.unzip = void 0;
var zlib_1 = __importDefault(require("zlib"));
var es6_promisify_1 = require("es6-promisify");
var unzip_pako_1 = require("./unzip-pako");
Object.defineProperty(exports, "pakoUnzip", { enumerable: true, get: function () { return unzip_pako_1.pakoUnzip; } });
Object.defineProperty(exports, "unzipChunk", { enumerable: true, get: function () { return unzip_pako_1.unzipChunk; } });
Object.defineProperty(exports, "unzipChunkSlice", { enumerable: true, get: function () { return unzip_pako_1.unzipChunkSlice; } });
var gunzip = (0, es6_promisify_1.promisify)(zlib_1.default.gunzip);
// in node, just use the native unzipping with Z_SYNC_FLUSH
function nodeUnzip(input) {
    //@ts-ignore
    return gunzip(input, {
        finishFlush: (zlib_1.default.constants || zlib_1.default).Z_SYNC_FLUSH,
    });
}
exports.unzip = nodeUnzip;
exports.nodeUnzip = nodeUnzip;
