# Use the Repeat Browser

Instruction on how to navigate [WashU Repetitive Element Browser](https://repeatbrowser.org/)

<br />
<br /> 

## Main Page
### The Top Menu
![The Top Menu](https://wangftp.wustl.edu/~scheng/repeat_browser/documentation/top%20menu.png)
The top navigation menu (above) controls the three main parts of Repeat Browser functionality. From the left to right are the Homepage, Input Data, and Data Visualization. You can also get access to documentation and the links to Wang Lab and WashU Med.

### Homepage
![Home Page](https://wangftp.wustl.edu/~scheng/repeat_browser/documentation/home_page_main.png)
You can start exploring WashU Repeat Browser from the Homepage (above). The **Github repo** section provides the raw codes and the **Start Browsing** button will lead to **Input Data** page to start browsing. A hands-on detailed guide can be called via **Start Guide Tour** button where you can get an overview of how to fully utilize the Repeat Browser. 

<br />
<br /> 

## Input Data
Input Data page is the prerequisite for visualization at the next step.

### Selected Data
![Selected Data](https://wangftp.wustl.edu/~scheng/repeat_browser/documentation/InputData_selected_data.png)
All the selected datasets and repeats would be organized at this page, and some information would be provided (files' ID and target, subfamily information of repeats). A button to download **Session JSON file** (similar to the Session ID file in WashU Epigenome Browser) is also provided to achieve easy downloading and efficient sharing.

### Files Selection
![Files Selection](https://wangftp.wustl.edu/~scheng/repeat_browser/documentation/InputData_files_selection.png)
Pre-processed public datasets can be archieved at this page, the datasets are categorised by Assay, Biosample, and Species (above). You can also check the selected files or start the hands-on tour through the right panel of this page. Lastly, the default files and repeats can also be reset by the button below.<br/>
Detailed files information can be further explored by clicking the numbers to call the form below. The information of datasets, including ID, Assasy, Biosample Type, Experiment ID, Organism, Tissue, and Status, can be explored. The green and red colors of Status column represent available and selected respectively.
![Files Selection Detailed Form](https://wangftp.wustl.edu/~scheng/repeat_browser/documentation/InputData_files_selection_detailed.png)

### Upload
![Files Selection](https://wangftp.wustl.edu/~scheng/repeat_browser/documentation/InputData_Upload.png)
Upload page provides two functions, **Upload Session Json** and **Zarr File URL**.<br/>
You can upload the JSON session file that is downloaded or shared to achieve quick exploration between collaborators.<br/>
You can also upload the personalized Zarr file generated by our [data processing pipeline](https://github.com/jamesc99/WashU_RepeatBrowser) (detailed information can be checked via this link)

### Repeats Selection
![Repeats Selection](https://wangftp.wustl.edu/~scheng/repeat_browser/documentation/InputData_repeat_selection.png)
Repeats page (above) provides sunburst plot for easy selection of repeats on class, family and subfamily levels. User can use **Repeats Search** bar to filter specific TE subfamily.<br/>
Sunburst plot can be played with via the GIF demo below. **Single click** on the panel would lead to downstream level (class to family to subfamily), and the first **double click** on the panel will select specific class, family, or subfamily, in which case the colorful panel would become gray. The second **double click** on the panel would undo the selction operation which would lead the panel to colorful one. In contrast, the **single click** on the central white region would back to the upstream level, while the **double click** on the central white region would back to the class level by default.
![Sunburnt](https://github.com/jamesc99/WashU_RepeatBrowser/blob/main/Documentation/repeats%20selection%20sunburnt.gif)

<br />
<br /> 

## Data Visualization
Data Visualization page is the main function of WashU Repeat Browser.

### Heatmap View
![Heatmap View](https://wangftp.wustl.edu/~scheng/repeat_browser/documentation/DataVisual_Heatmap.png)
Heatmap View (above) is the first plot inbeded in WashU Repeat Browser. It provides the functions where users can cluster columns(repeats) or rows(datasets), or both, as well as re-order the heatmap by any of the methods, alphabetic order, rank by sum, or rank by variance. You can also set the opacity and the number of rows you want to show by the side bar. The different colors at the edge of the matrix denote the different class/family (for the column) and different datasets (for the row).<br/>

:bangbang:(demo gif needed)

You can also switch to the CAGE-seq heatmap through the Assay Type drop down list.<br/>
Heatmap view of enrichment scores of datasets on TE subfamilies. Users can click a specific cell where each cell denotes ratio of average score or RPKM as per below. 
![Heatmap RPKM Calculation](https://wangftp.wustl.edu/~scheng/repeat_browser/documentation/Heatmap_RPKM.png)


### Consensus View
![Jump to Consensus View](https://github.com/jamesc99/WashU_RepeatBrowser/tree/main/Documentation/Heatmap_jumpto_Consensus.gif)
:bangbang:Users can click a specific cell and jump to the Consensus View panel (GIF above).

![Consensus View](https://wangftp.wustl.edu/~scheng/repeat_browser/documentation/DataVisual_ConsensusVIew.png)
Consensus View panel (above) displays epigenomic data anchored on a TE consensus sequence. Both unique reads and multi-reads can be displayed.<br/>

You can explore different biological meanings with different assay types, like open chromatin regions with ATAC-seq data, gene expression level with CAGE-seq, and TF binding sites with ChIP-seq data. The specific TE subfamily on different datasets can also be compared through the **Add Track** function, and the All reads and Uniquely-mapped reads can also be explored. Besides, the genome coverage and raw consensus sequence are also provided for further exploration.
:bangbang:(demo gif needed)

### Genome View
:bangbang:Users can click the **Genome View** on the left panel or the right-buttom button to jump to the Genome View panel (GIF above).

![Genome View](https://wangftp.wustl.edu/~scheng/repeat_browser/documentation/DataVisual_GenomeView.png)
Genome view (above) is a birdâ€™s eye view of different locations of TE copies on the genome. A filter sets the threshold of enrichment scores of corresponding TE copies is provided.

You can get the distribution of the uniquely-mapped reads of specific TE subfamily on Genome View and different shades represent different RPKM values (the darker the higher), where you adjust the threshold via the side bar to only show selected reads. The specific TE subfamily on different datasets can also be compared through the **Add Genome Copies** function, where you can explore the expression differences among different datasets. 
:bangbang:(demo gif needed)

Furthermore, to enhance the utility of Repeat Browser, we also link the Repeat Browser with WashU Epigenome Browser where users can choose view the interested individual TE copies on the WashU Epigenome Browser, using the **Region set View** function by which we can display multiple genomic locations side by side,optionally with customizable upstream and  downstream flanking regions (below).
![Region View Set](https://wangftp.wustl.edu/~scheng/repeat_browser/documentation/HeLa-S3_STAT1_MER41B_region_view_set.png)





