// smui/FieldTip/FieldTip.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".field-tip {\n  position: absolute;\n  max-width: 200px;\n  line-height: 1.5em;\n  font-size: 12px;\n  left: 0;\n  top: 0;\n}\n.field-tip .warning {\n  color: #C00;\n  min-width: 10em;\n}\n.field-tip .message {\n  color: #333;\n  min-width: 10em;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var params = {
    message: '',
    warning: '',
    for: '',
    position: 'absolute',
    show: false,
    placed: '',
    auto: false,
    tries: 0,
    width: 0,
    left: 0,
    top: 0,
    x: 0,
    y: 0
};

exports.default = {
    params: params,
    mixins: [_ui2.default],
    computed: {
        style: function style() {
            if (!this.show || !this.message && !this.warning) {
                return { display: 'none' };
            }
            var st = {
                position: this.position,
                left: this.x + 'px',
                top: this.y + 'px',
                display: 'block'
            };
            if (this.width) {
                st.width = this.width + 'px';
            }
            return st;
        }
    },
    ready: function ready() {
        this.init();
    },

    methods: {
        attachTo: function attachTo(el) {
            this.x = el.offsetWidth + el.offsetLeft + +this.left;
            this.y = el.offsetTop + +this.top;
        },
        init: function init() {
            var $vm = this;
            var target = null;
            if (!$vm.$el) {
                return;
            }
            if (this.placed) {
                this.message = this.placed.replace(/\[([a-z\/]+)\]/g, function (match, seg) {
                    return '<' + seg + '>';
                });
            }

            if (this.for) {
                target = this.$parent.$el.querySelector(this.for);
            }

            var fork = 0;
            var check = function check() {
                $vm.attachTo(target);
                clearTimeout(fork);
                fork = setTimeout(function () {
                    $vm.warning = target.validationMessage;
                    $vm.message = target.message || $vm.palced;
                }, 100);
            };

            if (!target && $vm.$el.parentElement) {
                $vm.$el.parentElement.querySelector('.field');
            }

            if (!target) {
                target = $vm.$el.previousElementSibling;
            }
            if (target) {
                $vm.attachTo(target);
                target.addEventListener('change', check);

                target.showMessage = check;
                $vm.show = true;
            }
            if (!target && this.tries < 3) {
                setTimeout(function () {
                    $vm.init();
                }, 1000);
                this.tries++;
                return;
            }
            if (this.auto) {
                check();
            }
            this.target = target;
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"field-tip",style:(_vm.style)},[(_vm.warning)?_c('div',{staticClass:"warning"},[_vm._v(_vm._s(_vm.warning))]):_vm._e(),_vm._v(" "),(_vm.message && !_vm.warning)?_c('div',{staticClass:"message",domProps:{"innerHTML":_vm._s(_vm.message)}}):_vm._e()])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-3060b0d7", __vue__options__)
  } else {
    hotAPI.reload("data-v-3060b0d7", __vue__options__)
  }
})()}
