// smui/StarRate/index.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".star-rate {\n  background-repeat: repeat-x;\n  height: 18px;\n  width: 70px;\n  text-align: left;\n  margin: 5px auto;\n  background-image: url('//image.uc.cn/s/uae/g/2p/material/images/star-empty.png');\n}\n.star-rate .rate {\n  height: 100%;\n  width: 50%;\n  background-repeat: repeat-x;\n  background-image: url('//image.uc.cn/s/uae/g/2p/material/images/star.png');\n  display: inline-block;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    props: ['params'],
    data: function data() {
        return (0, _assign2.default)({
            value: 0,
            max: 5
        }, this.params || {});
    },

    watch: {
        params: function params(val) {
            (0, _assign2.default)(this, val);
        }
    },
    methods: {
        getWidth: function getWidth() {
            var value = this.value,
                max = this.max;

            return value * 100 / max + '%';
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"star-rate"},[_c('div',{staticClass:"rate",style:({width: _vm.getWidth()})})])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-5a37099d", __vue__options__)
  } else {
    hotAPI.reload("data-v-5a37099d", __vue__options__)
  }
})()}
