// smui/SugBox/Index.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".sug-box-wrapper {\n  position: relative;\n  display: inline-block;\n  min-width: 120px;\n}\n.sug-box-wrapper input {\n  line-height: 28px;\n  min-width: 120px;\n  height: 28px;\n}\n.sug-box-wrapper .layer {\n  position: absolute;\n  min-width: 100%;\n  background: #FFF;\n  border: 1px solid #CCC;\n  top: 27px;\n  left: 0;\n  z-index: 9;\n  max-height: 300px;\n}\n.sug-box-wrapper .layer li {\n  cursor: pointer;\n  line-height: 25px;\n  padding: 5px;\n}\n.sug-box-wrapper .layer li:hover {\n  background: #DEDEDE;\n  color: #FFF;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var storage = window.localStorage;

var params = {
    value: '',
    fieldName: '',
    autoComplete: true,
    isInputing: true,
    showLayer: false
};

exports.default = {
    mixins: [_ui2.default],
    params: params,
    data: function data() {
        return (0, _assign2.default)({
            items: []
        }, this.params || {});
    },
    mounted: function mounted() {
        if (this.autoComplete && this.fieldName) {
            var cached = storage.getItem(this.storeKey);
            if (cached) {
                this.items = cached.split('\n');
            }
        }
    },

    computed: {
        storeKey: function storeKey() {
            return 'sug_box_' + this.fieldName;
        }
    },
    methods: {
        inputing: function inputing(e) {
            this.$emit('input-change', { value: this.value });
        },
        leave: function leave() {
            this.showLayer = false;
            if (this.value && this.items.indexOf(this.value) === -1) {
                this.items.push(this.value.trim());
                if (this.fieldName) {
                    storage.setItem(this.storeKey, this.items.join('\n'));
                }
            }
        },
        setItems: function setItems(items) {
            this.items = items;
        },
        setValue: function setValue(val) {
            this.value = val;
            this.showLayer = false;
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"sug-box-wrapper",on:{"mouseleave":_vm.leave}},[_c('input',{directives:[{name:"model",rawName:"v-model",value:(_vm.value),expression:"value"}],ref:"inputBox",attrs:{"type":"text","name":_vm.fieldName,"autocomplete":"off"},domProps:{"value":_vm._s(_vm.value)},on:{"input":[function($event){if($event.target.composing){ return; }_vm.value=$event.target.value},_vm.inputing],"click":function($event){_vm.showLayer = true}}}),_vm._v(" "),_c('ul',{directives:[{name:"show",rawName:"v-show",value:(_vm.showLayer),expression:"showLayer"}],staticClass:"layer"},_vm._l((_vm.items),function(item){return _c('li',{on:{"click":function($event){_vm.setValue(item)}}},[_vm._v(_vm._s(item))])}))])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-4f54e91e", __vue__options__)
  } else {
    hotAPI.reload("data-v-4f54e91e", __vue__options__)
  }
})()}
