// smui/Scrum/Scrum.js
var __vueify_style_dispose__ = require("vueify/lib/insert-css").insert(".scrum {\n  margin: 10px 0;\n}\n.scrum ul {\n  margin: 0;\n  padding: 0;\n}\n.scrum li {\n  list-style: none;\n  display: inline-block;\n  padding: 0;\n  padding-right: 15px;\n  color: #42b983;\n  cursor: pointer;\n}\n.scrum li:last-child {\n  color: #777;\n}\n.scrum li:last-child i {\n  display: none;\n}")
;(function(){
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ui = require('../base/ui');

var _ui2 = _interopRequireDefault(_ui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    mixins: [_ui2.default],
    params: {
        items: []
    },
    methods: {
        changeLink: function changeLink(item) {
            this.emit('link-change', item);
        }
    }
};
})()
if (module.exports.__esModule) module.exports = module.exports.default
var __vue__options__ = (typeof module.exports === "function"? module.exports.options: module.exports)
if (__vue__options__.functional) {console.error("[vueify] functional components are not supported and should be defined in plain js files using render functions.")}
__vue__options__.render = function render () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{staticClass:"scrum"},[_c('ul',_vm._l((_vm.items),function(item){return _c('li',[_c('a',{on:{"click":function($event){_vm.changeLink(item)}}},[_vm._v(_vm._s(item.text))]),_vm._v(" "),_c('i',{staticClass:"sm-arrow-right"})])}))])}
__vue__options__.staticRenderFns = []
if (module.hot) {(function () {  var hotAPI = require("vueify/node_modules/vue-hot-reload-api")
  hotAPI.install(require("vue"), true)
  if (!hotAPI.compatible) return
  module.hot.accept()
  module.hot.dispose(__vueify_style_dispose__)
  if (!module.hot.data) {
    hotAPI.createRecord("data-v-b8ba2d26", __vue__options__)
  } else {
    hotAPI.reload("data-v-b8ba2d26", __vue__options__)
  }
})()}
